{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BHv41D0vO1z7"
      },
      "source": [
        "#Necessary Imports"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "cyIQjowROumZ"
      },
      "outputs": [],
      "source": [
        "from ast import Pass\n",
        "import math\n",
        "from itertools import combinations\n",
        "from math import isclose\n",
        "import pandas as pd"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bxaaMFN8O-z9",
        "outputId": "de0674dd-d65d-4747-c4ee-9e37e95ff379"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: nltk in /usr/local/lib/python3.10/dist-packages (3.8.1)\n",
            "Requirement already satisfied: click in /usr/local/lib/python3.10/dist-packages (from nltk) (8.1.7)\n",
            "Requirement already satisfied: joblib in /usr/local/lib/python3.10/dist-packages (from nltk) (1.4.2)\n",
            "Requirement already satisfied: regex>=2021.8.3 in /usr/local/lib/python3.10/dist-packages (from nltk) (2024.5.15)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from nltk) (4.66.4)\n"
          ]
        }
      ],
      "source": [
        "pip install nltk"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jPE5DOxuPFuC",
        "outputId": "fe842cc4-a1ec-4d08-ead5-1f70c403de7e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "[nltk_data] Downloading package stopwords to /root/nltk_data...\n",
            "[nltk_data]   Package stopwords is already up-to-date!\n"
          ]
        }
      ],
      "source": [
        "import nltk\n",
        "nltk.download('stopwords')\n",
        "from nltk.corpus import wordnet\n",
        "from nltk.wsd import lesk\n",
        "from nltk.tokenize import word_tokenize\n",
        "from nltk.corpus import stopwords\n",
        "from nltk.stem import PorterStemmer\n",
        "from nltk import wordnet, pos_tag\n",
        "from nltk.corpus import wordnet as wn\n",
        "from sklearn.feature_extraction.text import CountVectorizer"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ypxa56hOgL3F",
        "outputId": "75b1d86d-d165-48d1-c642-814f7b992bb5"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "[nltk_data] Downloading package punkt to /root/nltk_data...\n",
            "[nltk_data]   Unzipping tokenizers/punkt.zip.\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "True"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "nltk.download('punkt')"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from ir_explain.pairwise.axioms import pairwise"
      ],
      "metadata": {
        "id": "QJZ8TfAskuR5"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from ir_explain.pairwise.explain_more import explain_more"
      ],
      "metadata": {
        "id": "OIr-0GM6lcOK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wm4IAHqe13Ke"
      },
      "source": [
        "Showcase some features of the ir_explain.pairwise framework example data:\n",
        "\n",
        "*   TREC-DL-2019\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SaOPRVsq1sxE",
        "outputId": "5dbb6150-2b17-47c4-9a0d-58ed28a634c1"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Collecting ir_datasets\n",
            "  Downloading ir_datasets-0.5.7-py3-none-any.whl (337 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m337.9/337.9 kB\u001b[0m \u001b[31m2.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: beautifulsoup4>=4.4.1 in /usr/local/lib/python3.10/dist-packages (from ir_datasets) (4.12.3)\n",
            "Collecting inscriptis>=2.2.0 (from ir_datasets)\n",
            "  Downloading inscriptis-2.5.0-py3-none-any.whl (45 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m45.4/45.4 kB\u001b[0m \u001b[31m4.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: lxml>=4.5.2 in /usr/local/lib/python3.10/dist-packages (from ir_datasets) (4.9.4)\n",
            "Requirement already satisfied: numpy>=1.18.1 in /usr/local/lib/python3.10/dist-packages (from ir_datasets) (1.25.2)\n",
            "Requirement already satisfied: pyyaml>=5.3.1 in /usr/local/lib/python3.10/dist-packages (from ir_datasets) (6.0.1)\n",
            "Requirement already satisfied: requests>=2.22.0 in /usr/local/lib/python3.10/dist-packages (from ir_datasets) (2.31.0)\n",
            "Requirement already satisfied: tqdm>=4.38.0 in /usr/local/lib/python3.10/dist-packages (from ir_datasets) (4.66.4)\n",
            "Collecting trec-car-tools>=2.5.4 (from ir_datasets)\n",
            "  Downloading trec_car_tools-2.6-py3-none-any.whl (8.4 kB)\n",
            "Collecting lz4>=3.1.10 (from ir_datasets)\n",
            "  Downloading lz4-4.3.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.3 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.3/1.3 MB\u001b[0m \u001b[31m18.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting warc3-wet>=0.2.3 (from ir_datasets)\n",
            "  Downloading warc3_wet-0.2.3-py3-none-any.whl (13 kB)\n",
            "Collecting warc3-wet-clueweb09>=0.2.5 (from ir_datasets)\n",
            "  Downloading warc3-wet-clueweb09-0.2.5.tar.gz (17 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting zlib-state>=0.1.3 (from ir_datasets)\n",
            "  Downloading zlib-state-0.1.6.tar.gz (9.5 kB)\n",
            "  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "  Installing backend dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting ijson>=3.1.3 (from ir_datasets)\n",
            "  Downloading ijson-3.2.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (111 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m111.8/111.8 kB\u001b[0m \u001b[31m12.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting unlzw3>=0.2.1 (from ir_datasets)\n",
            "  Downloading unlzw3-0.2.2-py3-none-any.whl (6.1 kB)\n",
            "Requirement already satisfied: soupsieve>1.2 in /usr/local/lib/python3.10/dist-packages (from beautifulsoup4>=4.4.1->ir_datasets) (2.5)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests>=2.22.0->ir_datasets) (3.3.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests>=2.22.0->ir_datasets) (3.7)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests>=2.22.0->ir_datasets) (2.0.7)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests>=2.22.0->ir_datasets) (2024.2.2)\n",
            "Collecting cbor>=1.0.0 (from trec-car-tools>=2.5.4->ir_datasets)\n",
            "  Downloading cbor-1.0.0.tar.gz (20 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Building wheels for collected packages: warc3-wet-clueweb09, zlib-state, cbor\n",
            "  Building wheel for warc3-wet-clueweb09 (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for warc3-wet-clueweb09: filename=warc3_wet_clueweb09-0.2.5-py3-none-any.whl size=18919 sha256=564cc2feeae7864339bdfb90b0ca7b318b3ef5a4c874021f07d4466036fcfbd1\n",
            "  Stored in directory: /root/.cache/pip/wheels/1a/d7/91/7ffb991df87e62355d945745035470ba2616aa3d83a250b5f9\n",
            "  Building wheel for zlib-state (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for zlib-state: filename=zlib_state-0.1.6-cp310-cp310-linux_x86_64.whl size=21162 sha256=f4a185e461116f2392914ef5b5642666b255edc0f2d792b873abaef9c4acb9a2\n",
            "  Stored in directory: /root/.cache/pip/wheels/32/72/7e/aff80f26e926b6e1fb08dfb52aba03c0e058f5e2258deb50a9\n",
            "  Building wheel for cbor (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for cbor: filename=cbor-1.0.0-cp310-cp310-linux_x86_64.whl size=53431 sha256=2a2fc544e1c7868430e0d97feef6b1313cdb08b05ee00cb9e3715d9cfe83baf3\n",
            "  Stored in directory: /root/.cache/pip/wheels/85/df/c9/b39e40eccaf76dbd218556639a6dc81562226f4c6a64902c85\n",
            "Successfully built warc3-wet-clueweb09 zlib-state cbor\n",
            "Installing collected packages: warc3-wet-clueweb09, warc3-wet, ijson, cbor, zlib-state, unlzw3, trec-car-tools, lz4, inscriptis, ir_datasets\n",
            "Successfully installed cbor-1.0.0 ijson-3.2.3 inscriptis-2.5.0 ir_datasets-0.5.7 lz4-4.3.3 trec-car-tools-2.6 unlzw3-0.2.2 warc3-wet-0.2.3 warc3-wet-clueweb09-0.2.5 zlib-state-0.1.6\n"
          ]
        }
      ],
      "source": [
        "!pip install ir_datasets"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "tFcrverTINUo"
      },
      "source": [
        "Get some query from the list to test and display the text of the query"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iE8HiH21cBZu",
        "outputId": "30fb824d-c909-4a2b-d3f0-66d210ec9213"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "[INFO] [starting] https://msmarco.z22.web.core.windows.net/msmarcoranking/msmarco-test2019-queries.tsv.gz\n",
            "[INFO] [finished] https://msmarco.z22.web.core.windows.net/msmarcoranking/msmarco-test2019-queries.tsv.gz: [00:00] [4.28kB] [16.2MB/s]\n",
            "                                                                                                                   "
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Query ID: 183378\n",
            "Query Text: exons definition biology\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "\r"
          ]
        }
      ],
      "source": [
        "dataset = ir_datasets.load(\"msmarco-document/trec-dl-2019\")\n",
        "query_id = \"183378\"\n",
        "\n",
        "query_text = None\n",
        "for query in dataset.queries_iter():\n",
        "    if query.query_id == query_id:\n",
        "        query_text = query.text\n",
        "        break\n",
        "if query_text is not None:\n",
        "    print(f\"Query ID: {query_id}\\nQuery Text: {query_text}\")\n",
        "else:\n",
        "    print(f\"No query found with ID: {query_id}\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "z68fy6cmIlzI"
      },
      "source": [
        "Get the top ranked documents for the given query"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "mDv9R4kCduj_",
        "outputId": "b39d3ba8-5283-4387-f198-9dc3e2ad84e9"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "[INFO] [starting] https://msmarco.z22.web.core.windows.net/msmarcoranking/msmarco-doctest2019-top100.gz\n",
            "[INFO] [finished] https://msmarco.z22.web.core.windows.net/msmarcoranking/msmarco-doctest2019-top100.gz: [00:00] [220kB] [1.10MB/s]\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Rank: 1, Doc ID: D1077802, Score: -6.03738\n",
            "Rank: 2, Doc ID: D2319972, Score: -6.19266\n",
            "Rank: 3, Doc ID: D3291537, Score: -6.2536\n",
            "Rank: 4, Doc ID: D359891, Score: -6.39307\n",
            "Rank: 5, Doc ID: D1806793, Score: -6.53852\n",
            "Rank: 6, Doc ID: D674120, Score: -6.61844\n",
            "Rank: 7, Doc ID: D3291538, Score: -6.66812\n",
            "Rank: 8, Doc ID: D1626054, Score: -6.6795\n",
            "Rank: 9, Doc ID: D3291534, Score: -6.68517\n",
            "Rank: 10, Doc ID: D1796583, Score: -6.70215\n",
            "Rank: 11, Doc ID: D3546110, Score: -6.7995\n",
            "Rank: 12, Doc ID: D906730, Score: -6.90626\n",
            "Rank: 13, Doc ID: D3291533, Score: -6.90913\n",
            "Rank: 14, Doc ID: D2646768, Score: -6.93343\n",
            "Rank: 15, Doc ID: D3555880, Score: -6.93369\n",
            "Rank: 16, Doc ID: D2790851, Score: -6.9459\n",
            "Rank: 17, Doc ID: D2319971, Score: -6.94791\n",
            "Rank: 18, Doc ID: D2657361, Score: -6.97759\n",
            "Rank: 19, Doc ID: D1012311, Score: -7.01531\n",
            "Rank: 20, Doc ID: D2108995, Score: -7.03303\n",
            "Rank: 21, Doc ID: D3380457, Score: -7.03347\n",
            "Rank: 22, Doc ID: D1806795, Score: -7.07862\n",
            "Rank: 23, Doc ID: D2970614, Score: -7.07915\n",
            "Rank: 24, Doc ID: D218234, Score: -7.08622\n",
            "Rank: 25, Doc ID: D2790852, Score: -7.11443\n",
            "Rank: 26, Doc ID: D1077806, Score: -7.15781\n",
            "Rank: 27, Doc ID: D3549517, Score: -7.16084\n",
            "Rank: 28, Doc ID: D1446709, Score: -7.20213\n",
            "Rank: 29, Doc ID: D74605, Score: -7.20381\n",
            "Rank: 30, Doc ID: D1349682, Score: -7.20409\n",
            "Rank: 31, Doc ID: D1779084, Score: -7.20416\n",
            "Rank: 32, Doc ID: D359893, Score: -7.2144\n",
            "Rank: 33, Doc ID: D3135852, Score: -7.22153\n",
            "Rank: 34, Doc ID: D2218493, Score: -7.2229\n",
            "Rank: 35, Doc ID: D674118, Score: -7.22465\n",
            "Rank: 36, Doc ID: D1436264, Score: -7.22465\n",
            "Rank: 37, Doc ID: D2646769, Score: -7.24324\n",
            "Rank: 38, Doc ID: D1578050, Score: -7.2554\n",
            "Rank: 39, Doc ID: D3394069, Score: -7.25614\n",
            "Rank: 40, Doc ID: D2319973, Score: -7.25671\n",
            "Rank: 41, Doc ID: D359896, Score: -7.26377\n",
            "Rank: 42, Doc ID: D2144833, Score: -7.28596\n",
            "Rank: 43, Doc ID: D2412278, Score: -7.3047\n",
            "Rank: 44, Doc ID: D1723687, Score: -7.30679\n",
            "Rank: 45, Doc ID: D1203069, Score: -7.31349\n",
            "Rank: 46, Doc ID: D922485, Score: -7.31351\n",
            "Rank: 47, Doc ID: D2902421, Score: -7.31434\n",
            "Rank: 48, Doc ID: D740424, Score: -7.3237\n",
            "Rank: 49, Doc ID: D297561, Score: -7.32425\n",
            "Rank: 50, Doc ID: D2805444, Score: -7.33659\n",
            "Rank: 51, Doc ID: D359892, Score: -7.34705\n",
            "Rank: 52, Doc ID: D2088207, Score: -7.34777\n",
            "Rank: 53, Doc ID: D1310908, Score: -7.34777\n",
            "Rank: 54, Doc ID: D359895, Score: -7.34927\n",
            "Rank: 55, Doc ID: D3337972, Score: -7.35208\n",
            "Rank: 56, Doc ID: D994262, Score: -7.3623\n",
            "Rank: 57, Doc ID: D1820803, Score: -7.37187\n",
            "Rank: 58, Doc ID: D630756, Score: -7.38447\n",
            "Rank: 59, Doc ID: D1077805, Score: -7.38919\n",
            "Rank: 60, Doc ID: D2004672, Score: -7.38934\n",
            "Rank: 61, Doc ID: D1229438, Score: -7.40724\n",
            "Rank: 62, Doc ID: D1143559, Score: -7.40782\n",
            "Rank: 63, Doc ID: D3480127, Score: -7.4207\n",
            "Rank: 64, Doc ID: D509044, Score: -7.42383\n",
            "Rank: 65, Doc ID: D334710, Score: -7.4253\n",
            "Rank: 66, Doc ID: D750935, Score: -7.42915\n",
            "Rank: 67, Doc ID: D383257, Score: -7.43177\n",
            "Rank: 68, Doc ID: D3232380, Score: -7.43414\n",
            "Rank: 69, Doc ID: D1330083, Score: -7.43945\n",
            "Rank: 70, Doc ID: D2958441, Score: -7.44947\n",
            "Rank: 71, Doc ID: D707807, Score: -7.45189\n",
            "Rank: 72, Doc ID: D2639761, Score: -7.46345\n",
            "Rank: 73, Doc ID: D731917, Score: -7.46669\n",
            "Rank: 74, Doc ID: D2797831, Score: -7.4692\n",
            "Rank: 75, Doc ID: D1077803, Score: -7.47256\n",
            "Rank: 76, Doc ID: D1628401, Score: -7.48941\n",
            "Rank: 77, Doc ID: D1024989, Score: -7.49589\n",
            "Rank: 78, Doc ID: D1442340, Score: -7.50105\n",
            "Rank: 79, Doc ID: D2904340, Score: -7.50709\n",
            "Rank: 80, Doc ID: D3188000, Score: -7.51495\n",
            "Rank: 81, Doc ID: D143210, Score: -7.51514\n",
            "Rank: 82, Doc ID: D2786081, Score: -7.51588\n",
            "Rank: 83, Doc ID: D1518712, Score: -7.52383\n",
            "Rank: 84, Doc ID: D143207, Score: -7.52629\n",
            "Rank: 85, Doc ID: D1871563, Score: -7.52818\n",
            "Rank: 86, Doc ID: D894961, Score: -7.53385\n",
            "Rank: 87, Doc ID: D214551, Score: -7.53547\n",
            "Rank: 88, Doc ID: D674119, Score: -7.53614\n",
            "Rank: 89, Doc ID: D2559864, Score: -7.54483\n",
            "Rank: 90, Doc ID: D1440407, Score: -7.55264\n",
            "Rank: 91, Doc ID: D2467575, Score: -7.55976\n",
            "Rank: 92, Doc ID: D645941, Score: -7.56113\n",
            "Rank: 93, Doc ID: D1285067, Score: -7.56425\n",
            "Rank: 94, Doc ID: D922486, Score: -7.56442\n",
            "Rank: 95, Doc ID: D1505279, Score: -7.56968\n",
            "Rank: 96, Doc ID: D1973953, Score: -7.57324\n",
            "Rank: 97, Doc ID: D886334, Score: -7.57786\n",
            "Rank: 98, Doc ID: D1349059, Score: -7.57923\n",
            "Rank: 99, Doc ID: D1578051, Score: -7.57929\n",
            "Rank: 100, Doc ID: D668156, Score: -7.5804\n"
          ]
        }
      ],
      "source": [
        "import ir_datasets\n",
        "\n",
        "dataset = ir_datasets.load(\"msmarco-document/trec-dl-2019\")\n",
        "\n",
        "target_query_id = \"183378\"\n",
        "\n",
        "target_query_docs = []\n",
        "\n",
        "for scoreddoc in dataset.scoreddocs_iter():\n",
        "    if scoreddoc.query_id == target_query_id:\n",
        "        target_query_docs.append((scoreddoc.doc_id, scoreddoc.score))\n",
        "i=1\n",
        "for doc_id, score in target_query_docs:\n",
        "    print(f\"Rank: {i}, Doc ID: {doc_id}, Score: {score}\")\n",
        "    i+=1\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Hw3B2f3rJSai"
      },
      "source": [
        "Lets take the top ranked document and test it against a lower ranked document"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "background_save": true,
          "base_uri": "https://localhost:8080/"
        },
        "id": "774xccnOPovq",
        "outputId": "e961ebfb-457c-4407-b606-ce88e5bca445"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "[INFO] If you have a local copy of https://msmarco.z22.web.core.windows.net/msmarcoranking/msmarco-docs.trec.gz, you can symlink it here to avoid downloading it again: /root/.ir_datasets/downloads/d4863e4f342982b51b9a8fc668b2d0c0\n",
            "[INFO] [starting] https://msmarco.z22.web.core.windows.net/msmarcoranking/msmarco-docs.trec.gz\n",
            "[INFO] [finished] https://msmarco.z22.web.core.windows.net/msmarcoranking/msmarco-docs.trec.gz: [02:24] [8.50GB] [58.8MB/s]\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Body of Document ID D1077802:\n",
            "Definition references  (+ images)Wikipedia:\n",
            "Exonany nucleotide sequence encoded by a gene that remains present within the final mature RNA product of that gene after introns have been removed by RNA splicing.\n",
            "The term exon refers to both the DNA sequence within a gene and to the corresponding...  (43 of 1099 words)en.wikipedia.org/wiki/Exon  [cite]Exon (disambiguation)may refer to: Exon, a region of DNA that is represented in the mature form of RNA | Exoniensis or Exon., the Post-Nominal Letters for alumni / degrees from the University of Exeter | Exon can also refer to the signature of the Bishop of Exeter | J...  (45 of 126 words, 7 definitions)en.wikipedia.org/wiki/Exon (disambiguation)  [cite]Collins Dictionary:exon | exonic  [adjective, derived] [British] one of the four officers who command the Yeomen of the Guard | any segment of a discontinuous gene the segments of which are separated by ...  (26 of 299 words, 2 definitions, 1 usage example, pronunciation)www.collinsdictionary.com/dictionary/english/exon  [cite]Oxford Dictionary:exon  [entry 1] | exonic  [derived]Syllabification: (ex·on) | [biochemistry] | a segment of a DNA or RNA molecule containing information coding for a protein or peptide sequence.\n",
            "(20 of 52 words, 3 definitions, pronunciation)oxforddictionaries.com/definition/american english/exon  [cite]exon  [entry 1, biochemistry] | exonic  [derived]a segment of a DNA or RNA molecule containing information coding for a protein or peptide sequence  (17 of 47 words, pronunciation)oxforddictionaries.com/definition/english/exon  [cite]exon  [entry 2]Entry from World dictionary | [British] | each of the four officers acting as commanders of the Yeomen of the Guard.\n",
            "(19 of 68 words, 3 definitions, pronunciation)oxforddictionaries.com/definition/american english/exon--2  [cite]exon  [entry 2, British]each of the four officers acting as commanders of the Yeomen of the Guard  (14 of 63 words, pronunciation)oxforddictionaries.com/definition/english/exon--2  [cite]Random House Dictionary:exon | exempt  [synonym, sense-specific] [British] one of four yeomen of the guard who act as commanding officers in the absence of higher authority.\n",
            "| \"exon\" : [genetics] : any portion of an ...  (27 of 54 words, 2 definitions, pronunciation)dictionary.infoplease.com/exon  [cite]Merriam-Webster:exon | exonic  [adjective, related]a polynucleotide sequence in a nucleic acid that codes information for protein synthesis and that is copied and spliced together with other such ...  (23 of 48 words, pronunciations)www.merriam-webster.com/dictionary/exon  [cite]Wiktionary:exon | exons  [plural]An officer of the Queen's Body Guard of the Yeomen of the Guard.\n",
            "| [genetics] A region of a transcribed gene present in the final functional RNA molecule.\n",
            "(27 of 29 words, 2 definitions)en.wiktionary.org/wiki/exon  [cite]American Heritage Dictionary:exon | exonic  [adjective, related]A sequence of DNA that codes information for protein synthesis that is transcribed to messenger RNA.\n",
            "(16 of 25 words, pronunciation)www.yourdictionary.com/exon  [cite]New World Dictionary:exona sequence in the genetic code that supplies the information for protein formation  (13 of 15 words, pronunciation)www.yourdictionary.com/exon  [cite]Britannica Encyclopedia:\n",
            "DNA  [chemical compound] | deoxyribonucleic acidorganic chemical of complex molecular structure that is found in all prokaryotic and eukaryotic cells and in many viruses.\n",
            "DNA codes genetic ...  (22 of 442 words, 16 images)www.britannica.com/EBchecked/topic/167063/DNA  [cite]Columbia Encyclopedia:nucleic acid | nucleic acids | codon | DNA | repressor | RNAany of a group of organic substances found in the chromosomes of living cells and viruses that play a central role in the storage and replication of ...  (27 of 259 words)www.infoplease.com/ce6/sci/A0836145.html  [cite]Macmillan British Dictionary:\n",
            "DNA  [uncountable, biology, abbreviation] | deoxyribonucleic acid | dnaa chemical substance that contains genetic information and is found in all living cells and some viruses  (17 of 47 words, pronunciation)www.macmillandictionary.com/dictionary/british/DNA  [cite]Cambridge Dictionary:\n",
            "DNA  [specialized, abbreviation] | deoxyribonucleic acidthe chemical at the centre of the cells of living things, which controls the structure and purpose of each cell and carries the genetic information ...  (25 of 35 words, pronunciation)dictionary.cambridge.org/dictionary/british/dna  [cite]Encarta Dictionary:exon | exons  [plural]a discontinuous sequence of DNA that codes for protein synthesis and carries the genetic code for the final messenger RNA molecule.\n",
            "intron  (22 of 46 words, pronunciation)encarta.msn.com/dictionary 1861609793/definition.html  [offline]\n",
            "\n",
            "\n"
          ]
        }
      ],
      "source": [
        "target_doc_id = \"D1077802\"\n",
        "\n",
        "doc1 = \"\"\n",
        "\n",
        "for doc in dataset.docs_iter():\n",
        "    if doc.doc_id == target_doc_id:\n",
        "        doc1 = doc.body\n",
        "        break\n",
        "\n",
        "if doc1 is not None:\n",
        "    print(f\"Body of Document ID {target_doc_id}:\\n{doc1}\")\n",
        "else:\n",
        "    print(f\"No document found with ID {target_doc_id}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "background_save": true
        },
        "id": "NsCCCOk1dP-Q",
        "outputId": "3adc56da-f953-4271-a491-d804b38d1bf6"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Body of Document ID D1806793:\n",
            "Skip to main page content\n",
            "HOME ABOUT ARCHIVE SUBMIT SUBSCRIBE ADVERTISE AUTHOR INFO CONTACT HELPSearch for Keyword: GO Advanced Search\n",
            "What is a gene, post-ENCODE?\n",
            "History and updated definition\n",
            "Mark B. Gerstein 1, 2, 3, 9, Can Bruce 2, 4, Joel S. Rozowsky 2, Deyou Zheng 2,Jiang Du 3, Jan O. Korbel 2, 5, Olof Emanuelsson 6, Zhengdong D. Zhang 2,Sherman Weissman 7, and Michael Snyder 2, 8+ Author Affiliations1Program in Computational Biology & Bioinformatics, Yale University, New Haven, Connecticut 06511, USA;2Molecular Biophysics & Biochemistry Department, Yale University, New Haven, Connecticut 06511, USA;3Computer Science Department, Yale University, New Haven, Connecticut 06511, USA;4Center for Medical Informatics, Yale University, New Haven, Connecticut 06511, USA;5European Molecular Biology Laboratory, 69117 Heidelberg, Germany;6Stockholm Bioinformatics Center, Albanova University Center, Stockholm University, SE-10691 Stockholm, Sweden;7Genetics Department, Yale University, New Haven, Connecticut 06511, USA;8Molecular, Cellular, & Developmental Biology Department, Yale University, New Haven, Connecticut 06511, USANext Section\n",
            "Abstract\n",
            "While sequencing of the human genome surprised us with how many protein-coding genes there are, it did not fundamentally change  our perspective on what a gene is.\n",
            "In contrast, the complex patterns of dispersed regulation and pervasive transcription uncovered  by the ENCODE project, together with non-genic conservation and the abundance of noncoding RNA genes, have challenged the  notion of the gene.\n",
            "To illustrate this, we review the evolution of operational definitions of a gene over the past century—from  the abstract elements of heredity of Mendel and Morgan to the present-day ORFs enumerated in the sequence databanks.\n",
            "We then  summarize the current ENCODE findings and provide a computational metaphor for the complexity.\n",
            "Finally, we propose a tentative  update to the definition of a gene: A gene is a union of genomic sequences encoding a coherent set of potentially overlapping  functional products.\n",
            "Our definition sidesteps the complexities of regulation and transcription by removing the former altogether  from the definition and arguing that final, functional gene products (rather than intermediate transcripts) should be used  to group together entities associated with a single gene.\n",
            "It also manifests how integral the concept of biological function  is in defining genes.\n",
            "Previous Section\n",
            "Next Section\n",
            "Introduction\n",
            "The classical view of a gene as a discrete element in the genome has been shaken by ENCODEThe ENCODE consortium recently completed its characterization of 1% of the human genome by various high-throughput experimental  and computational techniques designed to characterize functional elements ( The ENCODE Project Consortium 2007 ).\n",
            "This project represents a major milestone in the characterization of the human genome, and the current findings show a  striking picture of complex molecular activity.\n",
            "While the landmark human genome sequencing surprised many with the small number  (relative to simpler organisms) of protein-coding genes that sequence annotators could identify (∼21,000, according to the  latest estimate [see www.ensembl.org ]), ENCODE highlighted the number and complexity of the RNA transcripts that the genome produces.\n",
            "In this regard, ENCODE has  changed our view of “what is a gene” considerably more than the sequencing of the Haemophilus influenza and human genomes did ( Fleischmann et al.\n",
            "1995; Lander et al.\n",
            "2001; Venter et al.\n",
            "2001 ).\n",
            "The discrepancy between our previous protein-centric view of the gene and one that is revealed by the extensive transcriptional  activity of the genome prompts us to reconsider now what a gene is.\n",
            "Here, we review how the concept of the gene has changed  over the past century, summarize the current thinking based on the latest ENCODE findings, and propose a new updated gene  definition that takes these findings into account.\n",
            "Previous Section\n",
            "Next Section\n",
            "History of the gene, 1860 to just before ENCODEDefinition 1860s–1900s: Gene as a discrete unit of heredity\n",
            "The concept of the “gene” has evolved and become more complex since it was first proposed (see timeline in attached Figure  1, accompanying poster ).\n",
            "There are various definitions of the term, although common initial descriptions include the ability to determine a particular  characteristic of an organism and the heritability of this characteristic.\n",
            "In particular, the word gene was first used by Wilhelm Johannsen in 1909, based on the concept developed by Gregor Mendel in 1866 ( Mendel 1866 ).\n",
            "The word was a derivative of pangene, which was used by Hugo De Vries for entities involved in pangenesis, Darwin’s hypothetical mechanism of heredity ( Heimans 1962 ).\n",
            "Johnannsen called a gene the “special conditions, foundations and determiners which are present [in the gametes] in unique,  separate and thereby independent ways [by which] many characteristics of the organism are specified” ( Johannsen 1909, p. 124).\n",
            "The etymology of the term derives from the Greek genesis  (“birth”) or genos  (“origin”).\n",
            "The related word genetics was used by the geneticist William Bateson in 1905 ( http://www.jic.ac.uk/corporate/about/bateson.htm ).\n",
            "Mendel showed that when breeding plants, some traits such as height or flower color do not appear blended in their offspring—that  is, these traits are passed on as distinct, discrete entities ( Mendel 1866 ).\n",
            "His work also demonstrated that variations in traits were caused by variations in inheritable factors (or, in today’s terminology,  phenotype is caused by genotype).\n",
            "It was only after Mendel’s work was repeated and rediscovered by Carl Correns, Erich von  Tschermak-Seysenegg, and Hugo De Vries in 1900 that further work on the nature of the unit of inheritance truly began ( Tschermak 1900; Vries 1900; Rheinberger 1995 ).\n",
            "Definition 1910s: Gene as a distinct locus\n",
            "In the next major development, the American geneticist Thomas Hunt Morgan and his students were studying the segregation of  mutations in Drosophila melanogaster.\n",
            "They were able to explain their data with a model that genes are arranged linearly, and their ability to cross-over is proportional  to the distance that separated them.\n",
            "The first genetic map was created in 1913 ( Sturtevant 1913 ), and Morgan and his students published The Mechanism of Mendelian Inheritance in 1915 ( Morgan et al.\n",
            "1915 ).\n",
            "To the early geneticists, a gene was an abstract entity whose existence was reflected in the way phenotypes were transmitted  between generations.\n",
            "The methodology used by early geneticists involved mutations and recombination, so the gene was essentially  a locus whose size was determined by mutations that inactivated (or activated) a trait of interest and by the size of the  recombining regions.\n",
            "The fact that genetic linkage corresponded to physical locations on chromosomes was shown later, in 1929,  by Barbara Mc\n",
            "Clintock, in her cytogenetic studies on maize ( Mc\n",
            "Clintock 1929 ).\n",
            "Definition 1940s: Gene as a blueprint for a protein\n",
            "Beadle and Tatum (1941), who studied Neurospora metabolism, discovered that mutations in genes could cause defects in steps in metabolic pathways.\n",
            "This was stated as the  “one gene, one enzyme” view, which later became “one gene, one polypeptide.” In this viewpoint, the gene is being implicitly  considered as the information behind the individual molecules in a biochemical pathway.\n",
            "This view became progressively more  explicit and mechanistic in later decades.\n",
            "Definition 1950s: Gene as a physical molecule\n",
            "The fact that heredity has a physical, molecular basis was demonstrated by the observation that X rays could cause mutations  ( Muller 1927 ).\n",
            "Griffith’s (1928) demonstration that something in virulent but dead Pneumococcus strains could be taken up by live nonvirulent Pneumococcus and transform them into virulent bacteria was further evidence in this direction.\n",
            "It was later shown that this substance  could be destroyed by the enzyme DNase ( Avery et al.\n",
            "1944 ).\n",
            "In 1955, Hershey and Chase established that the substance actually transmitted by bacteriophage to their progeny is DNA  and not protein ( Hershey and Chase 1955 ).\n",
            "Moreover, the idea that a gene’s product is a diffusible substance underlies the complementation test that was used to  define genes in the early years of bacteriology.\n",
            "A practical view of the gene was that of the cistron, a region of DNA defined  by mutations that in trans could not genetically complement each other ( Benzer 1955 ).\n",
            "Definition 1960s: Gene as transcribed code\n",
            "It was the solution of the three-dimensional structure of DNA by Watson and Crick in 1953 ( Watson and Crick 1953) that explained how DNA could function as the molecule of heredity.\n",
            "Base pairing explained how genetic information could  be copied, and the existence of two strands explained how occasional errors in replication could lead to a mutation in one  of the daughter copies of the DNA molecule.\n",
            "From the 1960s on, molecular biology developed at a rapid pace.\n",
            "The RNA transcript of the protein-coding sequences was translated  using the genetic code (solved in 1965 by Nirenberg et al.\n",
            "[1965 and Söll et al.\n",
            "[1965) into an amino acid sequence.\n",
            "Francis Crick (1958) summarized the flow of information in gene expression as from nucleic acid to protein (the beginnings of the “Central Dogma”).\n",
            "However, there were some immediate exceptions to this: It was known that some genes code not for protein but for functional  RNA molecules such as r\n",
            "RNA and t\n",
            "RNA.\n",
            "In addition, in RNA viruses the gene is made of RNA.\n",
            "The molecular view of the gene that  developed through the 1960s can be summarized in general terms to be a code residing on nucleic acid that gives rise to a  functional product.\n",
            "Definition 1970s–1980s: Gene as open reading frame (ORF) sequence pattern\n",
            "The development of cloning and sequencing techniques in the 1970s, combined with knowledge of the genetic code, revolutionized  the field of molecular biology by providing a wealth of information on how genes are organized and expressed.\n",
            "The first gene  to be sequenced was from the bacteriophage MS2, which was also the first organism to be fully sequenced ( Fiers et al.\n",
            "1971, 1976 ).\n",
            "The parallel development of computational tools led to algorithms for the identification of genes based on their sequence  characteristics (e.g., for review, see Rogic et al.\n",
            "2001 ).\n",
            "In many cases, a DNA sequence could be used to infer structure and function for the gene and its products.\n",
            "This situation  created a new concept of the “nominal gene,” which is defined by its predicted sequence rather than as a genetic locus responsible  for a phenotype ( Griffiths and Stotz 2006 ).\n",
            "The identification of most genes in sequenced genomes is based either on their similarity to other known genes, or the  statistically significant signature of a protein-coding sequence.\n",
            "In many cases, the gene effectively became identified as  an annotated ORF in the genome ( Doolittle 1986 ).\n",
            "Definition 1990s–2000s: Annotated genomic entity, enumerated in the databanks (current view, pre-ENCODE)The current definition of a gene used by scientific organizations that annotate genomes still relies on the sequence view.\n",
            "Thus, a gene was defined by the Human Genome Nomenclature Organization as “a DNA segment that contributes to phenotype/function.\n",
            "In the absence of demonstrated function a gene may be characterized by sequence, transcription or homology” ( Wain et al.\n",
            "2002 ).\n",
            "Recently, the Sequence Ontology Consortium reportedly called the gene a “locatable region of genomic sequence, corresponding  to a unit of inheritance, which is associated with regulatory regions, transcribed regions and/or other functional sequence  regions” ( Pearson 2006 ).\n",
            "The sequencing of first the Haemophilus influenza genome and then the human genome ( Fleischmann et al.\n",
            "1995; Lander et al.\n",
            "2001; Venter et al.\n",
            "2001) led to an explosion in the amount of sequence that definitions such as the above could be applied to.\n",
            "In fact, there was  a huge popular interest in counting the number of genes in various organisms.\n",
            "This interest was crystallized originally by  Gene Sweepstake’s wager on the number of genes in the human genome, which received extensive media coverage ( Wade 2003 ).\n",
            "It has been pointed out that these enumerations overemphasize traditional, protein-coding genes.\n",
            "In particular, when the number  of genes present in the human genome was reported in 2003, it was acknowledged that too little was known about RNA-coding  genes, such that the given number was that of protein-coding genes.\n",
            "The Ensembl view of the gene was specifically summarized  in the rules of the Gene Sweepstake as follows: “alternatively spliced transcripts all belong to the same gene, even if the  proteins that are produced are different.” ( http://web.archive.org/web/20050627080719/www.ensembl.org/Genesweep/ ).\n",
            "A current computational metaphor: Genes as “subroutines” in the genomic operating system\n",
            "Given that counting genes in the genome is such a large-scale computational endeavor and that genes fundamentally deal with  information processing, the lexicon of computer science naturally has been increasingly applied to describing them.\n",
            "In particular,  people in the computational biology community have used the description of a formal language to describe the structure of  genes in very much the same way that grammars are used to describe computer programs—with a precise syntax of upstream regulation,  exons, and introns ( Searls 1997, 2001, 2002 ).\n",
            "Moreover, one metaphor that is increasingly popular for describing genes is to think of them in terms of subroutines in  a huge operating system (OS).\n",
            "That is, insofar as the nucleotides of the genome are put together into a code that is executed  through the process of transcription and translation, the genome can be thought of as an operating system for a living being.\n",
            "Genes are then individual subroutines in this overall system that are repetitively called in the process of transcription.\n",
            "Previous Section\n",
            "Next Section\n",
            "Problematic issues with the current definition of a gene\n",
            "There are a number of problematic aspects of the current definition of a gene, as applied to the human genome, which we discuss  below.\n",
            "Several additional complications are summarized in Table 1 .\n",
            "View this table:\n",
            "In this window In a new window\n",
            "Table 1.\n",
            "Phenomena complicating the concept of the gene1.\n",
            "Gene regulation\n",
            "Jacob and Monod (1961), in their study of the lac operon of Escherichia coli, provided a paradigm for the mechanism of regulation of the gene: It consisted of a region of DNA consisting of sequences  coding for one or more proteins, a “promoter” sequence for the binding of RNA polymerase, and an “operator” sequence to which  regulatory genes bind.\n",
            "Later, other sequences were found to exist that could affect practically every aspect of gene regulation  from transcription to m\n",
            "RNA degradation and post-translational modification.\n",
            "Such sequences could reside within the coding  sequence as well as in the flanking regions, and in the case of enhancers and related elements, very far away from the coding  sequence.\n",
            "Although functionally required for the expression of the gene product, regulatory elements, especially the distant  ones, made the concept of the gene as a compact genetic locus problematic.\n",
            "Regulation is integral to many current definitions of the gene.\n",
            "In particular, one current textbook definition of a gene in  molecular terms is the entire nucleic acid sequence that is necessary for the synthesis of a functional polypeptide (or RNA)  ( Lodish et al.\n",
            "2000 ).\n",
            "If that implies appropriately regulated synthesis, the DNA sequences in a gene would include not only those coding for  the pre-m\n",
            "RNA and its flanking control regions, but also enhancers.\n",
            "Moreover, many enhancers are distant along the DNA sequence,  although they are actually quite close due to three-dimensional chromatin structure.2.\n",
            "Overlapping and spliced genes\n",
            "Overlapping\n",
            "As genes, m\n",
            "RNAs, and eventually complete genomes were sequenced, the simple operon model turned out to be applicable only  to genes of prokaryotes and their phages.\n",
            "Eukaryotes were different in many respects, including genetic organization and information  flow.\n",
            "The model of genes as hereditary units that are nonoverlapping and continuous was shown to be incorrect by the precise  mapping of the coding sequences of genes.\n",
            "In fact, some genes have been found to overlap one another, sharing the same DNA  sequence in a different reading frame or on the opposite strand.\n",
            "The discontinuous structure of genes potentially allows one  gene to be completely contained inside another one’s intron, or one gene to overlap with another on the same strand without  sharing any exons or regulatory elements.\n",
            "Splicing\n",
            "Splicing was discovered in 1977 ( Berget et al.\n",
            "1977; Chow et al.\n",
            "1977; Gelinas and Roberts 1977 ).\n",
            "It soon became clear that the gene was not a simple unit of heredity or function, but rather a series of exons, coding  for, in some cases, discrete protein domains, and separated by long noncoding stretches called introns.\n",
            "With alternative splicing,  one genetic locus could code for multiple different m\n",
            "RNA transcripts.\n",
            "This discovery complicated the concept of the gene radically.\n",
            "For instance, in the sequencing of the genome, Celera defined a gene as “a locus of co-transcribed exons” ( Venter et al.\n",
            "2001 ), and Ensembl’s Gene Sweepstake Web page originally defined a gene as “a set of connected transcripts,” where “connected”  meant sharing one exon ( http://web.archive.org/web/20050428090317/www.ensembl.org/Genesweep ).\n",
            "The latter definition implies that a group of transcripts may share a set of exons, but no one exon is common to all of  them.\n",
            "Trans- splicing\n",
            "The phenomenon of trans -splicing (ligation of two separate m\n",
            "RNA molecules) further complicated our understanding ( Blumenthal 2005 ).\n",
            "There are examples of transcripts from the same gene, or the opposite DNA strand, or even another chromosome, being joined  before being spliced.\n",
            "Clearly, the classical concept of the gene as “a locus” no longer applies for these gene products whose  DNA sequences are widely separated across the genome.\n",
            "Finally, a number of recent studies have highlighted a phenomenon dubbed  tandem chimerism, where two consecutive genes are transcribed into a single RNA ( Akiva et al.\n",
            "2006; Parra et al.\n",
            "2006 ).\n",
            "The translation (after splicing) of such RNAs can lead to a new, fused protein, having parts from both original proteins.3.\n",
            "Parasitic and mobile genes\n",
            "A challenge to our concept of the gene has been that of the selfish or parasitic gene.\n",
            "The idea first proposed by Richard  Dawkins is that the unit of evolution is not the organism but the gene ( Dawkins 1976 ).\n",
            "Organisms are just tools that genes use to replicate themselves.\n",
            "Dawkins’ concept of the optimon (or selecton) is a unit  of DNA that survives recombination for enough generations to be selected for together.\n",
            "The term parasitic certainly appears appropriate for transposons, whose only function is to replicate themselves and which  do not provide any obvious benefit to the organism.\n",
            "Transposons can change their locations in addition to copying themselves  by excision, recombination, or reverse transcription.\n",
            "They were first discovered in the 1930s in maize and were later found  to exist in all branches of life, including humans ( Mc\n",
            "Clintock 1948 ).\n",
            "Transposons have altered our view of the gene by demonstrating that a gene is not fixed in its location.4.\n",
            "The large amount of “junk DNA” under selection\n",
            "The “ORF sequence pattern concept” of the gene as it existed from the 1980s onward made it clear that there were large expanses  of nongenic elements in eukaryotic genomes, particularly the human genome.\n",
            "In the absence of knowledge of a function for these  regions, it was proposed that they lacked a function by some who used the label of “junk DNA” ( Ohno 1972 ).\n",
            "This was underscored by the subsequent sequencing of the human genome, where it was shown that only 1.2% of the DNA bases  code for exons ( Lander et al.\n",
            "2001; Venter et al.\n",
            "2001 ).\n",
            "However, some early pilot functional genomics experiments on chromosomes 21 and 22 indicated that appreciable amounts of  the supposedly junk DNA were transcribed ( Kapranov et al.\n",
            "2002; Rinn et al.\n",
            "2003 ).\n",
            "Moreover, comparison of the human, dog, mouse, and other vertebrate genomes showed that a large fraction of these was conserved,  with ∼5% under negative selection since the divergence of these species ( Waterston et al.\n",
            "2002; Lindblad-Toh et al.\n",
            "2005 ).\n",
            "Previous Section\n",
            "Next Section\n",
            "The modern ENCODE view of dispersed genome activity\n",
            "As described above, before the advent of the ENCODE project, there were a number of aspects of genes that were very complicated,  but much of this complexity was in some sense swept under the rug and did not really affect the fundamental definition of  a gene.\n",
            "The experience of the ENCODE project, particularly the mapping of transcriptional activity and regulation using tiling  arrays, has extended these puzzling and confusing aspects of genes, bringing them to the forefront, where one has to grapple  more directly with them in relation to the definition of what a gene is.\n",
            "What the ENCODE experiments show: Lattices of long transcripts and dispersed regulation\n",
            "Unannotated transcription\n",
            "A first finding from the ENCODE consortium that has reproduced earlier results ( Bertone et al.\n",
            "2004; Cheng et al.\n",
            "2005) is that a vast amount of DNA, not annotated as known genes, is transcribed into RNA ( The ENCODE Project Consortium 2007 ).\n",
            "These novel transcribed regions are usually called TARs (i.e., transcriptionally active regions) and transfrags.\n",
            "While  the majority of the genome appears to be transcribed at the level of primary transcripts, only about half of the processed  (spliced) transcription detected across all the cell lines and conditions mapped is currently annotated as genes.\n",
            "Unannotated and alternative TSSs\n",
            "A second observation is that there are a large number of unannotated transcription start sites (TSSs) identified by either  sequencing of the 5′ end of transcribed m\n",
            "RNAs or the mapping of promoter-associated transcription factors via Ch\n",
            "IP–chip or  Ch\n",
            "IP–PET ( The ENCODE Project Consortium 2007 ).\n",
            "Furthermore, the consortium found that many known protein genes have alternative TSSs that are sometimes >100 kb upstream  of the annotated transcription start site.\n",
            "In particular, Denoeud et al.\n",
            "(2007) performed 5′ rapid amplification of c\n",
            "DNA ends (RACE) on all 399 well-characterized protein-coding loci contained in the ENCODE  regions.\n",
            "The RACE primer was selected from a 5′ exon that was shared among most annotated transcripts from each locus, and  the RACE products were hybridized to arrays and mapped.\n",
            "They found that more than half of the loci had an alternative transcription  start site upstream of the known site in at least one of the 12 tissues tested.\n",
            "Some of these distal TSSs used the promoter  of an entirely different gene locus (i.e., they share the same transcription start site).\n",
            "The significance of this discovery  is that the alternative TSS for some of these transcripts started two or three gene loci upstream of the locus from which  the RACE primer was selected.\n",
            "Thus, some alternative isoforms are transcripts that span multiple gene loci.\n",
            "(A cartoon schematic  is shown in Fig.\n",
            "2.)\n",
            "Many of the alternative isoforms code for the same protein differing only in their 5′ untranslated regions (UTRs).\n",
            "View larger version:\n",
            "In this page In a new window\n",
            "Download as Power\n",
            "Point Slide\n",
            "Figure 2.\n",
            "Biological complexity revealed by ENCODE.\n",
            "( A) Representation of a typical genomic region portraying the complexity of transcripts in the genome.\n",
            "( Top) DNA sequence with annotated exons of genes (black rectangles) and novel TARs (hollow rectangles).\n",
            "( Bottom) The various transcripts that arise from the region from both the forward and reverse strands.\n",
            "(Dashed lines) Spliced-out  introns.\n",
            "Conventional gene annotation would account for only a portion of the transcripts coming from the four genes in the  region (indicated).\n",
            "Data from the ENCODE project reveal that many transcripts are present that span across multiple gene loci,  some using distal 5′ transcription start sites.\n",
            "( B) Representation of the various regulatory sequences identified for a target gene.\n",
            "For Gene 1 we show all the component transcripts,  including many novel isoforms, in addition to all the sequences identified to regulate Gene 1 (gray circles).\n",
            "We observe that  some of the enhancer sequences are actually promoters for novel splice isoforms.\n",
            "Additionally, some of the regulatory sequences  for Gene 1 might actually be closer to another gene, and the target would be misidentified if chosen purely based on proximity.\n",
            "More alternative splicing\n",
            "Taking these findings into account, the Havana team at the Sanger Institute produced the well-curated GENCODE annotation ( Harrow et al.\n",
            "2006 ).\n",
            "They have not found that the number of known protein-coding gene loci has increased significantly over time.\n",
            "Conversely,  the number of annotated alternative isoforms per locus has increased.\n",
            "(The GENCODE annotation currently contains on average  5.4 transcripts per locus).\n",
            "Thus, while part of the large amount of new, unannotated transcription could correspond to entirely  new protein-coding gene loci, most of it is likely to correspond to segments of unannotated alternatively spliced transcripts  involving known gene loci or to entirely novel noncoding RNAs.\n",
            "Dispersed regulation\n",
            "As schematized in Figure 2B, the ENCODE project has provided evidence for dispersed regulation spread throughout the genome ( The ENCODE Project Consortium 2007 ).\n",
            "Moreover, the regulatory sites for a given gene are not necessarily directly upstream of it and can, in fact, be located  far away on the chromosome, closer to another gene.\n",
            "While the binding of many transcription factors appears to blanket the  entire genome, it is not arranged according to simple random expectations and tends to be clumped into regulatory rich “forests”  and poor “deserts” ( Zhang et al.\n",
            "2007 ).\n",
            "Moreover, it appears that some of regulatory elements may actually themselves be transcribed.\n",
            "In a conventional and concise  gene model, a DNA element (e.g., promoter, enhancer, and insulator) regulating gene expression is not transcribed and thus  is not part of a gene’s transcript.\n",
            "However, many early studies have discovered in specific cases that regulatory elements  can reside in transcribed regions, such as the lac operator ( Jacob and Monod 1961 ), an enhancer for regulating the beta-globin gene ( Tuan et al.\n",
            "1989 ), and the DNA binding site of YY1 factor ( Shi et al.\n",
            "1991 ).\n",
            "The ENCODE project and other recent Ch\n",
            "IP–chip experiments have provided large-scale evidence that the concise gene model  may be too simple, and many regulatory elements actually reside within the first exon, introns, or the entire body of a gene  ( Cawley et al.\n",
            "2004; Euskirchen et al.\n",
            "2004; Kim et al.\n",
            "2005; The ENCODE Project Consortium 2007; Zhang et al.\n",
            "2007 ).\n",
            "Genic versus intergenic: Is there a distinction?\n",
            "Overall, the ENCODE experiments have revealed a rich tapestry of transcription involving alternative splicing, covering the  genome in a complex lattice of transcripts.\n",
            "According to traditional definitions, genes are unitary regions of DNA sequence,  separated from each other.\n",
            "ENCODE reveals that if one attempts to define a gene on the basis of shared overlapping transcripts,  then many annotated distinct gene loci coalesce into bigger genomic regions.\n",
            "One obvious implication of the ENCODE results  is that there is less of a distinction to be made between genic and intergenic regions.\n",
            "Genes now appear to extend into what  was once called intergenic space, with newly discovered transcripts originating from additional regulatory sites.\n",
            "Moreover,  there is much activity between annotated genes in the intergenic space.\n",
            "Two well-characterized sources can contribute to this,  transcribed non-protein-coding RNAs (nc\n",
            "RNAs) and transcribed pseudogenes, and an appreciable fraction of these transcribed  elements are under evolutionary constraint.\n",
            "A number of these transcribed pseudogenes and nc\n",
            "RNA genes are, in fact, located  within introns of protein-coding genes.\n",
            "One cannot simply ignore these components within introns because some of them may  influence the expression of their host genes, either directly or indirectly.\n",
            "Noncoding RNAs\n",
            "The roles of nc\n",
            "RNA genes are quite diverse, including gene regulation (e.g., mi\n",
            "RNAs), RNA processing (e.g., sno\n",
            "RNAs), and  protein synthesis (t\n",
            "RNAs and r\n",
            "RNA) ( Eddy 2001; Mattick and Makunin 2006 ).\n",
            "Due to the lack of codons and thus open reading frames, nc\n",
            "RNA genes are hard to identify, and thus probably only a fraction  of the functional nc\n",
            "RNAs in humans is known to date, with the exception of the ones with the strongest evolutionary and/or  structural constraints, which can be identified computationally through RNA folding and coevolution analyses (e.g., mi\n",
            "RNAs  that display characteristic hairpin-shaped precursor structures, or nc\n",
            "RNAs in ribonucleoprotein complexes that in combination  with peptides form specific secondary structures) ( Washietl et al.\n",
            "2005, 2007; Pedersen et al.\n",
            "2006 ).\n",
            "However, the example of the 17-kb large XIST gene involved in dosage compensation shows that functional nc\n",
            "RNAs can expand significantly beyond constrained, computationally  identifiable regions ( Chureau et al.\n",
            "2002; Duret et al.\n",
            "2006 ).\n",
            "It is also possible that the RNA products themselves do not have a function, but rather reflect or are important for a particular  cellular process.\n",
            "For example, transcription of a regulatory region might be important for chromatin accessibility for transcription  factor binding or for DNA replication.\n",
            "Such transcription has been found in the locus control region (LCR) of the beta-globin  locus, and polymerase activity has been suggested to be important for DNA replication in E. coli.\n",
            "Alternatively, transcription might reflect nonspecific activity of a particular region, for example, the recruitment of  polymerase to regulatory sites.\n",
            "In either of these scenarios, the transcripts themselves would lack a function and be unlikely  to be conserved.\n",
            "Pseudogenes\n",
            "Pseudogenes are yet another group of “mysterious” genomic components that are often found in introns of genes or in intergenic  space ( Torrents et al.\n",
            "2003; Zhang et al.\n",
            "2003 ).\n",
            "They are derived from functional genes (through retrotransposition or duplication) but have lost the original functions  of their parental genes ( Balakirev and Ayala 2003 ).\n",
            "Sometimes swinging between dead and alive, pseudogenes can influence the structure and function of the human genome.\n",
            "Their  prevalence (as many as protein-coding genes) and their close similarity to functional genes have already confounded gene annotation.\n",
            "Recently, it has also been found that a significant fraction (up to 20%) of them are transcriptionally alive, suggesting that  care has to be taken when using expression as evidence for locating genes ( Yano et al.\n",
            "2004; Harrison et al.\n",
            "2005; Zheng et al.\n",
            "2005, 2007; Frith et al.\n",
            "2006 ).\n",
            "Indeed, some of the novel TARs can be attributed to pseudogene transcription ( Bertone et al.\n",
            "2004; Zheng et al.\n",
            "2005 ).\n",
            "In a few surprising cases, a pseudogene RNA or at least a piece of it was found to be spliced with the transcript of its  neighboring gene to form a gene–pseudogene chimeric transcript.\n",
            "These findings add one extra layer of complexity to establishing  the precise structure of a gene locus.\n",
            "Furthermore, functional pseudogene transcripts have also been discovered in eukaryotic  cells, such as the neurons of the snail Lymnaea stagnalis  ( Korneev et al.\n",
            "1999 ).\n",
            "Also, interestingly, the human XIST gene mentioned above actually arises from the dead body of a pseudogene ( Duret et al.\n",
            "2006 ).\n",
            "Pseudogene transcription and the blurring boundary between genes and pseudogenes ( Zheng and Gerstein 2007) emphasizes once more that the functional nature of many novel TARs needs to be resolved by future biochemical or genetic  experiments (for review, see Gingeras 2007 ).\n",
            "Constrained elements\n",
            "The noncoding intergenic regions contain a large fraction of functional elements identified by examining evolutionary changes  across multiple species and within the human population.\n",
            "The ENCODE project observed that only 40% of the evolutionarily constrained  bases were within protein-coding exons or their associated untranslated regions ( The ENCODE Project Consortium 2007 ).\n",
            "The resolution of constrained elements identified by multispecies analysis in the ENCODE project is very high, identifying  sequences as small as 8 bases (with a median of 19 bases) ( The ENCODE Project Consortium 2007 ).\n",
            "This suggests that protein-coding loci can be viewed as a cluster of small constrained elements dispersed in a sea of unconstrained  sequences.\n",
            "Approximately another 20% of the constrained elements overlap with experimentally annotated regulatory regions.\n",
            "Therefore, a similar fraction of constrained elements (40% in terms of bases) is located in protein-coding regions as unannotated  noncoding regions (100% – 40% coding – 20% regulatory regions), suggesting that the latter may be as functionally important  as the former.\n",
            "The ENCODE computational metaphor: Genes as “loosely coded” routines\n",
            "The new ENCODE perspective does not, of course, fit with the metaphor of the gene as a simple callable routine in a huge operating  system.\n",
            "In this new perspective, one enters a gene “routine” in many different ways in the framework of alternative splicing  and lattices of long transcripts.\n",
            "The execution of the genomic OS does not have as neat a quality as this idea of repetitive  calls to a discrete subroutine in a normal computer OS.\n",
            "However, the framework of describing the genome as executed code still  has some merit.\n",
            "That is, one can still understand gene transcription in terms of parallel threads of execution, with the caveat  that these threads do not follow canonical, modular subroutine structure.\n",
            "Rather, threads of execution are intertwined in  a rather “higgledy-piggledy” fashion, very much like what would be described as a sloppy, unstructured computer program code  with lots of GOTO statements zipping in and out of loops and other constructs.\n",
            "Previous Section\n",
            "Next Section\n",
            "The importance of gene models for interpreting the high-throughput experiment in ENCODEGiven the provocative findings of the ENCODE project, one wonders to what degree the interpretation of the high-throughput  experiments can be pushed.\n",
            "This interpretation is, in fact, very contingent on using gene models.\n",
            "Aspects of interpreting tiling array data\n",
            "A large part of the transcription data was generated using high-density tiling microarrays ( Emanuelsson et al.\n",
            "2007; Rozowsky et al.\n",
            "2007; The ENCODE Project Consortium 2007 ).\n",
            "The advantage of such arrays is that they probe the transcription in an unbiased and detailed way, with no preconceptions  as to where to look for activity.\n",
            "On the other hand, the output from a tiling array experiment can be noisy and needs careful  interpretation in order to allow the collection of a reliable set of transcribed regions.\n",
            "The amount of detected transcription  depends heavily on the thresholds used when calling transcribed regions and to some extent also on the segmentation algorithms  used to delineate transcribed regions from nontranscribed regions.\n",
            "Furthermore, since the ENCODE transcription mapping (and  other) experiments were carried out on many different tissues and cell lines, direct comparison between experiments is not  trivial, and the overlap between different transcription maps is sometimes quite low, partly due to the variable biological  features of the samples used in the experiments.\n",
            "The exact expected outcome of a transcription mapping experiment—the true transcription map—is, of course, unknown.\n",
            "Thus,  a crucial part of interpreting transcription mapping tiling array data is to understand how the signal is different from various  random expectations (null models).\n",
            "A naive way to achieve this goal is to randomize the raw data and then apply all the normalizing,  scoring, and segmentation schemes (with unchanged parameters) to get a “baseline” of transcription that is picked up using  supposedly meaningless data.\n",
            "But it is not clear that this is the best way to do it: GC-content, length distribution of the  (actual or expected) transcribed regions, dinucleotide composition, and other characteristics should also be factored into  the “baseline” transcription distribution.\n",
            "The expected outcome also depends on the biological sample used: tissue or cell  line, developmental stage, external stimuli, etc.\n",
            "Integration of transcription maps from different biological sources (tissues,  cell lines) provides greater confidence in the result.\n",
            "Genes as statistical models summarizing many experiments\n",
            "In the context of interpreting high-throughput experiments such as tiling arrays, the concept of a gene has an added practical  importance—as a statistical model to help interpret and provide concise summarization to potentially noisy experimental data.\n",
            "For example, the transcriptional tiling array experiments ultimately identify TARs/transfrags, which usually correspond to  the exons in genes.\n",
            "Therefore, the most appropriate gene models to be considered can be splicing graphs ( Heber et al.\n",
            "2002) representing exons as nodes and splicing events as directed edges.\n",
            "In order to build and adjust statistical models for experimental interpretation, other related biological knowledge (e.g.,  gene annotation and experimental validation data) needs to be combined with the high-throughput array data.\n",
            "For instance,  the transcriptional array data can identify isolated transcribed regions, and experimental validation such as RACE can provide  connectivity information.\n",
            "Using these data together, the statistical models can be better trained and can then be used to  analyze the rest of the high-throughput data that are not covered by the validation experiments.\n",
            "Different statistical models ( Karplus et al.\n",
            "1999; Bertone et al.\n",
            "2004; Schadt et al.\n",
            "2004; Gibbons et al.\n",
            "2005; Ji and Wong 2005; Li et al.\n",
            "2005; Du et al.\n",
            "2006; Marioni et al.\n",
            "2006) have been proposed to explain the generation of the tiling array data.\n",
            "As shown in Figure 3, these models can be trained using the tiling array data and other biological knowledge and then extrapolated to the whole  genome sequence to best segment it into functional elements.\n",
            "As more and more biological knowledge is accumulated, especially  via the experimental validation of predicted functional regions generated by the analysis procedure, we can expect that the  models will be better trained, thus leading to refined analysis results of these experiments.\n",
            "However, it will be impractical  to validate every single functional element identified by tiling experiments using RT-PCR or RACE.\n",
            "For each tiling array experiment,  perhaps only a medium-sized set of predicted functional regions will be validated experimentally.\n",
            "View larger version:\n",
            "In this page In a new window\n",
            "Download as Power\n",
            "Point Slide\n",
            "Figure 3.\n",
            "Training statistical gene models based on high-density oligonucleotide tiling microarray data.\n",
            "( A) Large-scale signal data from tiling array experiments can be used to train statistical models to score the hits, and a small/medium  proportion of these results can be further validated by experiments or other biological knowledge via runs of iterations and  optimizations.\n",
            "( B) Different strategies can be used to select genomic regions for validation; e.g., (1) select only the regions with high signals,  (2) select regions randomly, or (3) select those that have the maximum signal entropies, which usually contain “borders” of  high and low signals.\n",
            "One question worth asking is whether an optimal way of selection exists to best help in training the  statistical model.\n",
            "As shown in Figure 3, the regions for experimental validation can be picked using different strategies.\n",
            "It is obviously beneficial to pick these  regions in an optimal way so that the model trained based on these validation results can most accurately analyze the remainder  of the tiling array data.\n",
            "In a specific case, when analyzing tiling array data using a hidden Markov model ( Du et al.\n",
            "2006 ), if the validation regions are selected to achieve maximum signal entropy, the Max\n",
            "Entropy selection scheme, the resulting  gene segmentation model outperforms others.\n",
            "For transcriptional tiling arrays, Max\n",
            "Entropy will generally select regions containing  both exons and introns.\n",
            "Previous Section\n",
            "Next Section\n",
            "Toward an updated definition of a gene\n",
            "As we have described above, our knowledge of genes has evolved greatly over the past century.\n",
            "While our understanding has  grown, we have also uncovered an increasing number of problematic aspects with simple definitions of a gene ( Table 1 ).\n",
            "Splicing (including alternative splicing) and intergenic transcription are obviously some of the most problematic aspects.\n",
            "As shown in Figure 4, the frequency of mention of these terms in the biological literature has been increasing considerably.\n",
            "Thus, the stage was  set for the ENCODE project and the great complexity in transcriptional and regulatory apparatus that it highlighted.\n",
            "At this  point, it is not clear what to do: In the extreme, we could declare the concept of the gene dead and try to come up with something  completely new that fits all the data.\n",
            "However, it would be hard to do this with consistency.\n",
            "Here, we made a tentative attempt  at a compromise, devising updates and patches for the existing definition of a gene.\n",
            "View larger version:\n",
            "In this page In a new window\n",
            "Download as Power\n",
            "Point Slide\n",
            "Figure 4.\n",
            "Keyword analysis and complexity of genes.\n",
            "Using Google Scholar, a full-text search of scientific articles was performed for  the keywords “intron,” “alternative splicing,” and “intergenic transcription.” Slopes of curves indicate that in recent years  the frequency of mentioning of terms relating to the complexity of a gene has increased.\n",
            "(The Google Scholar search was limited  to articles in the following subject areas: “Biology, Life Sciences, and Environmental Science;” “Chemistry and Materials  Science;” “Medicine, Pharmacology, and Veterinary Science.”)Criteria to consider in updating the definition\n",
            "First, we consider several criteria to be important while coming up with an updated definition for a gene: (1) A new definition  must attempt to be backward compatible, in the sense that something that used to be called a gene should remain a gene.\n",
            "(2) It must be organism-independent; i.e., be as valid for bacteria as for a virus or a higher eukaryote.\n",
            "(3) It should be a statement of a simple idea, rather than listing various mechanisms and exceptions.\n",
            "(4) It should be practical enough so that one can readily enumerate genes and answer a question like “How many genes are there in the human genome?” (5) It should be compatible with other biological nomenclature that makes use of the idea of a digital gene.\n",
            "For instance, it should be consistent with term regulome, which represents  the complete set of regulatory interactions in an organism.\n",
            "A proposed updated definition\n",
            "There are three aspects to the definition that we will list below, before providing the succinct definition:\n",
            "A gene is a genomic sequence (DNA or RNA) directly encoding functional product molecules, either RNA or protein.\n",
            "In the case that there are several functional products sharing overlapping regions, one takes the union of all overlapping genomic sequences coding for them.\n",
            "This union must be coherent —i.e., done separately for final protein and RNA products—but does not require that all products necessarily share a common  subsequence.\n",
            "This can be concisely summarized as:\n",
            "The gene is a union of genomic sequences encoding a coherent set of potentially overlapping functional products.\n",
            "Figure 5 provides an example to illustrate the application of this definition.\n",
            "View larger version:\n",
            "In this page In a new window\n",
            "Download as Power\n",
            "Point Slide\n",
            "Figure 5.\n",
            "How the proposed definition of the gene can be applied to a sample case.\n",
            "A genomic region produces three primary transcripts.\n",
            "After alternative splicing, products of two of these encode five protein products, while the third encodes for a noncoding  RNA (nc\n",
            "RNA) product.\n",
            "The protein products are encoded by three clusters of DNA sequence segments (A, B, and C; D; and E).\n",
            "In the case of the three-segment cluster (A, B, C), each DNA sequence segment is shared by at least two of the products.\n",
            "Two  primary transcripts share a 5′ untranslated region, but their translated regions D and E do not overlap.\n",
            "There is also one  noncoding RNA product, and because its sequence is of RNA, not protein, the fact that it shares its genomic sequences (X and  Y) with the protein-coding genomic segments A and E does not make it a co-product of these protein-coding genes.\n",
            "In summary,  there are four genes in this region, and they are the sets of sequences shown inside the orange dashed lines: Gene 1 consists  of the sequence segments A, B, and C; gene 2 consists of D; gene 3 of E; and gene 4 of X and Y.\n",
            "In the diagram, for clarity,  the exonic and protein sequences A and E have been lined up vertically, so the dashed lines for the spliced transcripts and  functional products indicate connectivity between the proteins sequences (ovals) and RNA sequences (boxes).\n",
            "(Solid boxes on  transcripts) Untranslated sequences, (open boxes) translated sequences.\n",
            "Aspects and implications of the definition\n",
            "There are important implications of this definition.\n",
            "Collapsing in simple cases\n",
            "In simple cases where the gene is not discontinuous or there are no overlapping products, our definition collapses to the  classical version of being a DNA sequence that codes for a protein or RNA product.\n",
            "Projecting down in taking union\n",
            "In our proposed definition of a gene, different functional products of the same class (protein or RNA) that overlap in their  usage of the primary DNA sequence are combined into the same gene.\n",
            "This overlap is done by projecting the sequence of the  final product (either amino acid or RNA sequence) down onto the original genomic sequence from which it was derived.\n",
            "One could  in principle overlap the sequences of the final products (“projecting up”); however, since annotation of genes is done for  genomic DNA, we think our choice is the most consistent with current practice.\n",
            "An obvious point that should still be stated  is that, when looking at genomic products with common sequence segments, mere sequence identity is not enough; the products  have to be encoded directly from the same genomic region.\n",
            "Thus, paralogous proteins may share sequence blocks, but DNA sequences  coding for them reside in separate locations in the genome, and so they would not constitute one gene.\n",
            "Frameshifted exons\n",
            "There are cases, such as that of the CDKN2A (formerly INK4a/ARF) tumor suppressor gene (e.g., Quelle et al.\n",
            "1995 ), when a pre-m\n",
            "RNA can be alternatively spliced to generate an m\n",
            "RNA with a frameshift in the protein sequence.\n",
            "Thus, although  the two m\n",
            "RNAs have coding sequences in common, the protein products may be completely different.\n",
            "This rather unusual case  brings up the question of how exactly sequence identity is to be handled when taking the union of sequence segments that are  shared among protein products.\n",
            "If one considers the sequence of the protein products, there are two unrelated proteins, so  there must be two genes with overlapping sequence sets.\n",
            "If one “projects” the sequence of the protein products back to the  DNA sequence that encoded them (as described above), then there are two sequence sets with common elements, so there is one  gene.\n",
            "The fact that these two proteins’ sequences are simultaneously constrained, such that a mutation in one of them would  simultaneously affect the other one, suggests that this situation is not akin to that of two unrelated protein-coding genes.\n",
            "For this reason, generalizing from this special case, we favor the method of taking the union of the sequence segments, not  of the products, but of the DNA sequences that code for the product sequences.\n",
            "Regulatory regions not included\n",
            "Although regulatory regions are important for gene expression, we suggest that they should not be considered in deciding whether  multiple products belong to the same gene.\n",
            "This aspect of the definition results from our concept of the bacterial operon.\n",
            "The fact that genes in an operon share an operator and promoter region has traditionally not been considered to imply that  their protein products are alternative products of a single gene.\n",
            "Consequently, in higher eukaryotes, two transcripts that  originate from the same transcription start site (sharing the same promoter and regulatory elements) but do not share any  sequence elements in their final products (e.g., because of alternative splicing) would not be products of the same gene.\n",
            "A similar logic would apply to multiple transcripts sharing a common but distant enhancer or insulator.\n",
            "Regulation is simply  too complex to be folded into the definition of a gene, and there is obviously a many-to-many (rather than one-to-one) relationship  between regulatory regions and genes.\n",
            "Final products, not transcript clusters\n",
            "As the updated definition emphasizes the final products of a gene, it disregards intermediate products originating from a  genomic region that may happen to overlap.\n",
            "For example, an intronic transcript clearly shares sequences with an overlapping  larger transcript, but this fact is irrelevant when we conclude that the two products share no sequence blocks.\n",
            "This concept  can be generalized to other types of discontinuous genes, such as rearranged genes (e.g., in the immunoglobulin gene locus,  the C segment is common to all protein products encoded from it), or trans -spliced transcripts (where one pre-m\n",
            "RNA can be spliced to a number of other pre-m\n",
            "RNAs before further processing and translation).\n",
            "This implies that the number of genes in the human genome is going to increase significantly when the survey of the human  transcriptome is completed.\n",
            "In light of the large amount of intertwined transcripts that were identified by the ENCODE consortium,  if we tried to cluster entire transcripts together to form overlapping transcript clusters (a potential alternate definition  of a gene), then we would find that large segments of chromosomes would coalesce into these clusters.\n",
            "This alternate definition  of a gene would result in far fewer “genes,” and would be of limited utility.\n",
            "Alternative splicing\n",
            "In relation to alternatively spliced gene products, there is the possibility that no one coding exon is shared among all protein  products.\n",
            "In this case, it is understood that the union of these sequence segments defines the gene, as long as each exon  is shared among at least two members of this group of products.\n",
            "UTRs5′ and 3′ untranslated regions (UTRs) play important roles in translation, regulation, stability, and/or localization of m\n",
            "RNAs.\n",
            "When using a strict definition of regions encoding the final product of a protein-coding gene, these regions would no longer  be considered part of the gene, as is often the case in current usage.\n",
            "Moreover, protein-coding transcripts that share DNA  sequence only in their untranslated regions or introns would not be clustered together into a common gene.\n",
            "By removing UTRs  from the definition of a gene, one can avoid the problem of multiple 5′ and 3′ ends clouding the delineation of the gene and  also avoid a situation in which upstream or trans 5′ leader sequences are spliced onto a protein coding sequence.\n",
            "Moreover, it has been observed that most of the longer protein-coding  transcripts identified by ENCODE differ only in their UTRs, and thus our definition is quite transparent to this degree of  transcript complexity.\n",
            "Gene-associated regions\n",
            "As described above, regulatory and untranslated regions that play an important part in gene expression would no longer be  considered part of the gene.\n",
            "However, we would like to create a special “category” for them, by saying that they would be gene-associated.\n",
            "In this way, these regions still retain their important role in contributing to gene function.\n",
            "Moreover, their ability to  contribute to the expression of several genes can be recognized.\n",
            "This is particularly true for long-range elements such as  the beta-globin LCR, which contributes to the expression of several genes, and will likely be the case for many other enhancers  as their true gene targets are mapped.\n",
            "It can also be applied to untranslated regions that contribute to multiple gene loci,  such as the long spliced transcripts observed in the ENCODE region and trans -spliced exons.\n",
            "Disjointed sets of genomic sequence\n",
            "For clarity in the discussion, we refer to “DNA” when meaning genomic sequences in general.\n",
            "Our proposed definition is applicable  to all genomes, including that of RNA viruses.\n",
            "In complex cases, the gene turns out not to correspond to a discrete single  genetic locus, as sequences coding for its product (s) may be widely separated in the genome.\n",
            "In particular, because the gene  is a set of sequences shared among the products, there is no requirement of connectivity between these sequences and the sequences  that happen to connect them need not be part of the gene.\n",
            "Thus, members of a sequence can be on different strands of a chromosome  or even on separate chromosomes.\n",
            "This means that trans -spliced transcripts belong to one gene.\n",
            "Previous Section\n",
            "Next Section\n",
            "Conclusion: What is function?\n",
            "The classical view of a gene as a unit of hereditary information aligned along a chromosome, each coding for one protein,  has changed dramatically over the past century.\n",
            "For Morgan, genes on chromosomes were like beads on a string.\n",
            "The molecular  biology revolution changed this idea considerably.\n",
            "To quote Falk (1986), ‘‘.\n",
            ".\n",
            ".\n",
            "the gene is [.\n",
            ".\n",
            ".\n",
            "]neither discrete [.\n",
            ".\n",
            ".]\n",
            "nor continuous [.\n",
            ".\n",
            ".\n",
            "], nor does it have a constant location [.\n",
            ".\n",
            ".\n",
            "],  nor a clearcut function [.\n",
            ".\n",
            ".\n",
            "], not even constant sequences [.\n",
            ".\n",
            ".]\n",
            "nor definite borderlines.’’ And now the ENCODE project  has increased the complexity still further.\n",
            "What has not changed is that genotype determines phenotype, and at the molecular level, this means that DNA sequences determine  the sequences of functional molecules.\n",
            "In the simplest case, one DNA sequence still codes for one protein or RNA.\n",
            "But in the  most general case, we can have genes consisting of sequence modules that combine in multiple ways to generate products.\n",
            "By  focusing on the functional products of the genome, this definition sets a concrete standard in enumerating unambiguously the  number of genes it contains.\n",
            "An important aspect of our proposed definition is the requirement that the protein or RNA products must be functional for the purpose of assigning them to a particular gene.\n",
            "We believe this connects to the basic principle of genetics, that  genotype determines phenotype.\n",
            "At the molecular level, we assume that phenotype relates to biochemical function.\n",
            "Our intention  is to make our definition backwardly compatible with earlier concepts of the gene.\n",
            "This emphasis on functional products, of course, highlights the issue of what biological function actually is.\n",
            "With this,  we move the hard question from “what is a gene?” to “what is a function?”High-throughput biochemical and mutational assays will be needed to define function on a large scale ( Lan et al.\n",
            "2002, 2003 ).\n",
            "Hopefully, in most cases it will just be a matter of time until we acquire the experimental evidence that will establish  what most RNAs or proteins do.\n",
            "Until then we will have to use “place-holder” terms like TAR, or indicate our degree of confidence  in assuming function for a genomic product.\n",
            "We may also be able to infer functionality from the statistical properties of  the sequence (e.g., Ponjavic et al.\n",
            "2007 ).\n",
            "However, we probably will not be able to ever know the function of all molecules in the genome.\n",
            "It is conceivable that some  genomic products are just “noise,” i.e., results of evolutionarily neutral events that are tolerated by the organism (e.g., Tress et al.\n",
            "2007 ).\n",
            "Or, there may be a function that is shared by so many other genomic products that identifying function by mutational approaches  may be very difficult.\n",
            "While determining biological function may be difficult, proving lack of function is even harder (almost  impossible).\n",
            "Some sequence blocks in the genome are likely to keep their labels of “TAR of unknown function” indefinitely.\n",
            "If such regions happen to share sequences with functional genes, their boundaries (or rather, the membership of their sequence  set) will remain uncertain.\n",
            "Given that our definition of a gene relies so heavily on functional products, finalizing the number  of genes in our genome may take a long time.\n",
            "Previous Section\n",
            "Next Section\n",
            "Acknowledgments\n",
            "We thank the ENCODE consortium, and acknowledge the following funding sources: ENCODE grant #U01HG03156 from the National  Human Genome Research Institute (NHGRI)/National Institutes of Health (NIH); NIH Grant T15 LM07056 from the National Library  of Medicine (C.\n",
            "B., Z.\n",
            "D.\n",
            "Z.\n",
            "); and a Marie Curie Outgoing International Fellowship (J.\n",
            "O.\n",
            "K).\n",
            "Previous Section\n",
            "Next Section\n",
            "Footnotes↵ 9 Corresponding author.↵ 9 E-mail Mark.\n",
            "Gerstein@yale.edu; fax (360) 838-7861.\n",
            "Article is online at http://www.genome.org/cgi/doi/10.1101/gr.6339607Freely available online through the Genome Research Open Access option.\n",
            "Copyright © 2007, Cold Spring Harbor Laboratory Press\n",
            "Previous Section\n",
            "References↵Akiva, P., Toporik, A., Edelheit, S., Peretz, Y., Diber, A., Shemesh, R., Novik, A.,Sorek, R. ( 2006) Transcription-mediated gene fusion in the human genome.\n",
            "Genome Res.\n",
            "16: 30 – 36.\n",
            "Abstract / FREE Full Text↵Avery, O.\n",
            "T., Mac\n",
            "Leod, C.\n",
            "M., Mc\n",
            "Carty, M. ( 1944) Studies on the chemical nature of the substance inducing transformation of pneumococcal types.\n",
            "J. Exp.\n",
            "Med.79: 137 – 158.\n",
            "Abstract↵Balakirev, E.\n",
            "S., Ayala, F.\n",
            "J.\n",
            "( 2003) Pseudogenes: Are they “junk” or functional DNA?\n",
            "Annu.\n",
            "Rev.\n",
            "Genet.\n",
            "37: 123 – 151.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Beadle, G.\n",
            "W., Tatum, E.\n",
            "L. ( 1941) Genetic control of biochemical reactions in Neurospora.\n",
            "Proc.\n",
            "Natl.\n",
            "Acad.\n",
            "Sci.\n",
            "27: 499 – 506.\n",
            "FREE Full Text↵Benzer, S. ( 1955) Fine structure of a genetic region in bacteriophage.\n",
            "Proc.\n",
            "Natl.\n",
            "Acad.\n",
            "Sci.\n",
            "41: 344 – 354.\n",
            "FREE Full Text↵Berget, S.\n",
            "M., Moore, C., Sharp, P.\n",
            "A.\n",
            "( 1977) Spliced segments at the 5′ terminus of adenovirus 2 late m\n",
            "RNA.\n",
            "Proc.\n",
            "Natl.\n",
            "Acad.\n",
            "Sci.\n",
            "74: 3171 – 3175.\n",
            "Abstract / FREE Full Text↵Bertone, P., Stolc, V., Royce, T.\n",
            "E., Rozowsky, J.\n",
            "S., Urban, A.\n",
            "E., Zhu, X., Rinn, J.\n",
            "L.,Tongprasit, W., Samanta, M., Weissman, S., et al.\n",
            "( 2004) Global identification of human transcribed sequences with genome tiling arrays.\n",
            "Science306: 2242 – 2246.\n",
            "Abstract / FREE Full Text↵Blumenthal, T. ( 2005) Trans-splicing and operons.\n",
            "Worm\n",
            "Book (ed.\n",
            "The C. elegans Research Community)Worm\n",
            "Book, doi/10.1895/wormbook.1.5.1, http://www.wormbook.org.\n",
            "Borst, P. ( 1986) Discontinuous transcription and antigenic variation in trypanosomes.\n",
            "Annu.\n",
            "Rev.\n",
            "Biochem.\n",
            "55: 701 – 732.\n",
            "Cross\n",
            "Ref Medline\n",
            "Google Scholar↵Cawley, S., Bekiranov, S., Ng, H.\n",
            "H., Kapranov, P., Sekinger, E.\n",
            "A., Kampa, D.,Piccolboni, A., Sementchenko, V., Cheng, J., Williams, A.\n",
            "J., et al.\n",
            "( 2004) Unbiased mapping of transcription factor binding sites along human chromosomes 21 and 22 points to widespread regulation of  noncoding RNAs.\n",
            "Cell 116: 499 – 509.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Cheng, J., Kapranov, P., Drenkow, J., Dike, S., Brubaker, S., Patel, S., Long, J.,Stern, D., Tammana, H., Helt, G., et al.\n",
            "( 2005) Transcriptional maps of 10 human chromosomes at 5-nucleotide resolution.\n",
            "Science 308: 1149 – 1154.\n",
            "Abstract / FREE Full Text↵Chow, L.\n",
            "T., Gelinas, R.\n",
            "E., Broker, T.\n",
            "R., Roberts, R.\n",
            "J.\n",
            "( 1977) An amazing sequence arrangement at the 5′ ends of adenovirus 2 messenger RNA.\n",
            "Cell 12: 1 – 8.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Chureau, C., Prissette, M., Bourdet, A., Barbe, V., Cattolico, L., Jones, L.,Eggen, A., Avner, P., Duret, L. ( 2002) Comparative sequence analysis of the X-inactivation center region in mouse, human, and bovine.\n",
            "Genome Res.12: 894 – 908.\n",
            "Abstract / FREE Full Text\n",
            "Contreras, R., Rogiers, R., Van de, V.\n",
            "A., Fiers, W. ( 1977) Overlapping of the VP2-VP3 gene and the VP1 gene in the SV40 genome.\n",
            "Cell 12: 529 – 538.\n",
            "Cross\n",
            "Ref\n",
            "Medline Google Scholar↵Crick, F.\n",
            "H.\n",
            "C. ( 1958) On protein synthesis.\n",
            "Symp.\n",
            "Soc.\n",
            "Exp.\n",
            "Biol.\n",
            "XII: 138 – 163.\n",
            "Google Scholar↵Dawkins, R. ( 1976) The selfish gene  ( Oxford University Press, Oxford, UK ).\n",
            "Google Scholar↵Denoeud, F., Kapranov, P., Ucla, C., Frankish, A., Castelo, R., Drenkow, J.,Lagarde, J., Alioto, T., Manzano, C., Chrast, J., et al.\n",
            "( 2007) Prominent use of distal 5′ transcription start sites and discovery of a large number of additional exons in ENCODE regions.\n",
            "Genome Res.\n",
            "10.1101/gr5660607,  (this issue).\n",
            "Google Scholar\n",
            "Dobrovic, A., Gareau, J.\n",
            "L., Ouellette, G., Bradley, W.\n",
            "E. ( 1988) DNA methylation and genetic inactivation at thymidine kinase locus: Two different mechanisms for silencing autosomal genes.\n",
            "Somat.\n",
            "Cell Mol.\n",
            "Genet.\n",
            "14: 55 – 68.\n",
            "Cross\n",
            "Ref Medline\n",
            "Google Scholar↵Doolittle, R. ( 1986) Of URFs and ORFs: A primer on how to analyze derived amino acid sequences  ( University Science Books, Mill Valley, CA ).\n",
            "Google Scholar↵Du, J., Rozowsky, J.\n",
            "S., Korbel, J., Zhang, Z.\n",
            "D., Royce, T.\n",
            "E., Schultz, M.\n",
            "H.,Snyder, M., Gerstein, M. ( 2006) A supervised hidden Markov model framework for efficiently segmenting tiling array data in transcriptional and Ch\n",
            "IP–chip experiments:  Systematically incorporating validated biological knowledge.\n",
            "Bioinformatics 22: 3016 – 3024.\n",
            "Abstract / FREE Full Text↵Duret, L., Chureau, C., Samain, S., Weissenbach, J., Avner, P. ( 2006) The Xist RNA gene evolved in eutherians by pseudogenization of a protein-coding gene.\n",
            "Science 312: 1653 – 1655.\n",
            "Abstract / FREE Full Text\n",
            "Early, P., Huang, H., Davis, M., Calame, K., Hood, L. ( 1980) An immunoglobulin heavy chain variable region gene is generated from three segments of DNA: VH, D and JH.\n",
            "Cell 19: 981 – 992.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Eddy, S.\n",
            "R. ( 2001) Non-coding RNA genes and the modern RNA world.\n",
            "Nat.\n",
            "Rev.\n",
            "Genet.\n",
            "2: 919 – 929.\n",
            "Cross\n",
            "Ref Medline Google Scholar\n",
            "Eisen, H. ( 1988) RNA editing: Who’s on first?\n",
            "Cell 53: 331 – 332.\n",
            "Cross\n",
            "Ref Medline\n",
            "Google Scholar↵Emanuelsson, O., Nagalakshmi, U., Zheng, D., Rozowsky, J.\n",
            "S., Urban, A.\n",
            "E.,Du, J., Lian, Z., Stolc, V., Weissman, S., Snyder, M., et al.\n",
            "( 2007) Assessing the performance of different high-density tiling microarray strategies for mapping transcribed regions of the human  genome.\n",
            "Genome Res.\n",
            "10.1101/gr.5014606,  (this issue).\n",
            "Abstract / FREE Full Text↵The ENCODE Project Consortium, ( 2007) Identification and analysis of functional elements in 1% of the human genome by the ENCODE pilot project.\n",
            "Nature,  (in press).\n",
            "Google Scholar↵Euskirchen, G., Royce, T.\n",
            "E., Bertone, P., Martone, R., Rinn, J.\n",
            "L., Nelson, F.\n",
            "K.,Sayward, F., Luscombe, N.\n",
            "M., Miller, P., Gerstein, M., et al.\n",
            "( 2004) CREB binds to multiple loci on human chromosome 22.\n",
            "Mol.\n",
            "Cell.\n",
            "Biol.\n",
            "24: 3804 – 3814.\n",
            "Abstract / FREE Full Text↵Falk, R. ( 1986) What is a gene?\n",
            "Stud.\n",
            "Hist.\n",
            "Philos.\n",
            "Sci.\n",
            "17: 133 – 173.\n",
            "Cross\n",
            "Ref\n",
            "Medline Google Scholar↵Fiers, W., Contreras, R., De Wachter, R., Haegeman, G., Merregaert, J.,Jou, W.\n",
            "M., Vandenberghe, A.\n",
            "( 1971) Recent progress in the sequence determination of bacteriophage MS2 RNA.\n",
            "Biochimie 53: 495 – 506.\n",
            "Medline\n",
            "Google Scholar↵Fiers, W., Contreras, R., Duerinck, F., Haegeman, G., Iserentant, D.,Merregaert, J., Min\n",
            "Jou, W., Molemans, F., Raeymakers, A., Van den Berghe, A.,et al.\n",
            "( 1976) Complete nucleotide sequence of bacteriophage MS2 RNA: Primary and secondary structure of the replicase gene.\n",
            "Nature 260: 500 – 507.\n",
            "Cross\n",
            "Ref\n",
            "Medline Google Scholar↵Fleischmann, R.\n",
            "D., Adams, M.\n",
            "D., White, O., Clayton, R.\n",
            "A., Kirkness, E.\n",
            "F.,Kerlavage, A.\n",
            "R., Bult, C.\n",
            "J., Tomb, J.\n",
            "F., Dougherty, B.\n",
            "A., Merrick, J.\n",
            "M., et al.\n",
            "( 1995) Whole-genome random sequencing and assembly of Haemophilus influenzae Rd.\n",
            "Science 269: 496 – 512.\n",
            "Abstract / FREE Full Text↵Frith, M.\n",
            "C., Wilming, L.\n",
            "G., Forrest, A., Kawaji, H., Tan, S.\n",
            "L., Wahlestedt, C.,Bajic, V.\n",
            "B., Kai, C., Kawai, J., Carninci, P., et al.\n",
            "( 2006) Pseudo-messenger RNA: Phantoms of the transcriptome.\n",
            "PLo\n",
            "S Genet.\n",
            "2: e23.\n",
            "Cross\n",
            "Ref Medline\n",
            "Google Scholar↵Gelinas, R.\n",
            "E., Roberts, R.\n",
            "J.\n",
            "( 1977) One predominant 5′-undecanucleotide in adenovirus 2 late messenger RNAs.\n",
            "Cell 11: 533 – 544.\n",
            "Cross\n",
            "Ref Medline\n",
            "Google Scholar↵Gibbons, F.\n",
            "D., Proft, M., Struhl, K., Roth, F.\n",
            "P. ( 2005) Chipper: Discovering transcription-factor targets from chromatin immunoprecipitation microarrays using variance stabilization.\n",
            "Genome Biol.\n",
            "6: R96.\n",
            "Cross\n",
            "Ref Medline\n",
            "Google Scholar↵Gingeras, T. ( 2007) Origin of phenotypes: Genes and transcripts.\n",
            "Genome Res.10.1101/gr.625007,  (this issue).\n",
            "Google Scholar↵Griffith, F. ( 1928) The significance of pneumococcal types.\n",
            "J. Hyg.\n",
            "(Lond.\n",
            ")27: 113 – 159.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Griffiths, P.\n",
            "E., Stotz, K. ( 2006) Genes in the postgenomic era.\n",
            "Theor.\n",
            "Med.\n",
            "Bioeth.\n",
            "27: 499 – 521.\n",
            "Cross\n",
            "Ref Medline Google Scholar\n",
            "Handa, H., Bonnard, G., Grienenberger, J.\n",
            "M. ( 1996) The rapeseed mitochondrial gene encoding a homologue of the bacterial protein Ccl1 is divided into two independently transcribed  reading frames.\n",
            "Mol.\n",
            "Gen. Genet.\n",
            "252: 293 – 302.\n",
            "Google Scholar↵Harrison, P.\n",
            "M., Zheng, D., Zhang, Z., Carriero, N., Gerstein, M. ( 2005) Transcribed processed pseudogenes in the human genome: An intermediate form of expressed retrosequence lacking protein-coding  ability.\n",
            "Nucleic Acids Res.33: 2374 – 2383.\n",
            "Abstract / FREE Full Text↵Harrow, J., Denoeud, F., Frankish, A., Reymond, A., Chen, C.\n",
            "K., Chrast, J.,Lagarde, J., Gilbert, J.\n",
            "G., Storey, R., Swarbreck, D., et al.\n",
            "( 2006) GENCODE: Producing a reference annotation for ENCODE.\n",
            "Genome Biol.\n",
            "7 Suppl.\n",
            "1: S4.1 – S9.\n",
            "Google Scholar↵Heber, S., Alekseyev, M., Sze, S., Tang, H., Pevzner, P.\n",
            "A.\n",
            "( 2002) Splicing graphs and EST assembly problem.\n",
            "Bioinformatics 18: S181 – S188.\n",
            "Abstract↵Heimans, J.\n",
            "( 1962) Hugo de Vries and the gene concept.\n",
            "Am.\n",
            "Nat.\n",
            "96: 93 – 104.\n",
            "Cross\n",
            "Ref Google Scholar\n",
            "Henikoff, S., Keene, M.\n",
            "A., Fechtel, K., Fristrom, J.\n",
            "W. ( 1986) Gene within a gene: Nested Drosophila genes encode unrelated proteins on opposite DNA strands.\n",
            "Cell 44: 33 – 42.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Hershey, A.\n",
            "D., Chase, M. ( 1955) An upper limit to the protein content of the germinal substance of bacteriophage T2.\n",
            "Virology 1: 108 – 127.\n",
            "Cross\n",
            "Ref Medline\n",
            "Google Scholar\n",
            "Iafrate, A.\n",
            "J., Feuk, L., Rivera, M.\n",
            "N., Listewnik, M.\n",
            "L., Donahoe, P.\n",
            "K., Qi, Y.,Scherer, S.\n",
            "W., Lee, C. ( 2004) Detection of large-scale variation in the human genome.\n",
            "Nat.\n",
            "Genet.\n",
            "36: 949 – 951.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Jacob, F., Monod, J.\n",
            "( 1961) Genetic regulatory mechanisms in the synthesis of proteins.\n",
            "J. Mol.\n",
            "Biol.\n",
            "3: 318 – 356.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Ji, H., Wong, W.\n",
            "H. ( 2005) Tile\n",
            "Map: Create chromosomal map of tiling array hybridizations.\n",
            "Bioinformatics 21: 3629 – 3636.\n",
            "Abstract / FREE Full Text↵Johannsen, W. ( 1909) Elemente der exakten Erblichkeitslehre, Jena.\n",
            "Quoted by Nils Roll-Hansen (1989).\n",
            "The crucial experiment of Wilhelm Johannsen.\n",
            "Biol.\n",
            "Philos.\n",
            "4: 303 – 329.\n",
            "Cross\n",
            "Ref Google Scholar↵Kapranov, P., Cawley, S.\n",
            "E., Drenkow, J., Bekiranov, S., Strausber, R.\n",
            "L.,Fodor, S.\n",
            "P., Gingeras, T.\n",
            "R. ( 2002) Large-scale transcriptional activity in chromosomes 21 and 22.\n",
            "Science 296: 916 – 919.\n",
            "Abstract / FREE Full Text↵Karplus, K., Barrett, C., Cline, M., Diekhans, M., Grate, L., Hughey, R. ( 1999) Predicting protein structure using only sequence information.\n",
            "Proteins 37 (Suppl 3) 121 – 125.\n",
            "Google Scholar↵Kim, T.\n",
            "H., Barrera, L.\n",
            "O., Zheng, M., Qu, C., Singer, M.\n",
            "A., Richmond, T.\n",
            "A.,Wu, Y., Green, R.\n",
            "D., Ren, B.\n",
            "( 2005) A high-resolution map of active promoters in the human genome.\n",
            "Nature 436: 876 – 880.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Korneev, S.\n",
            "A., Park, J.\n",
            "H., O’Shea, M. ( 1999) Neuronal expression of neural nitric oxide synthase (n\n",
            "NOS) protein is suppressed by an antisense RNA transcribed from an NOS  pseudogene.\n",
            "J. Neurosci.\n",
            "19: 7711 – 7720.\n",
            "Abstract / FREE Full Text↵Lan, N., Jansen, R., Gerstein, M. ( 2002) Towards a systematic definition of protein function that scales to the genome level: Defining function in terms of interactions.\n",
            "Proc.\n",
            "IEEE 90: 1848 – 1858.\n",
            "Cross\n",
            "Ref Google Scholar↵Lan, N., Montelione, G.\n",
            "T., Gerstein, M. ( 2003) Ontologies for proteomics: Towards a systematic definition of structure and function that scales to the genome level.\n",
            "Curr.\n",
            "Opin.\n",
            "Chem.\n",
            "Biol.\n",
            "7: 44 – 54.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Lander, E.\n",
            "S., Linton, L.\n",
            "M., Birren, B., Nusbaum, C., Zody, M.\n",
            "C., Baldwin, J.,Devon, K., Dewar, K., Doyle, M., Fitz\n",
            "Hugh, W., et al.\n",
            "( 2001) Initial sequencing and analysis of the human genome.\n",
            "Nature 409: 860 – 921.\n",
            "Cross\n",
            "Ref Medline\n",
            "Google Scholar↵Li, W., Meyer, C.\n",
            "A., Liu, X.\n",
            "S. ( 2005) A hidden Markov model for analyzing Ch\n",
            "IP–chip experiments on genome tiling arrays and its application to p53 binding sequences.\n",
            "Bioinformatics 21: i274 – i282.\n",
            "Abstract↵Lindblad-Toh, K.\n",
            "C.\n",
            "M., Wade, T.\n",
            "S., Mikkelsen, E.\n",
            "K., Karlsson, D.\n",
            "B., Jaffe, M.,Kamal, M., Clamp, J.\n",
            "L., Chang, E.\n",
            "J., Kulbokas, M.\n",
            "C., 3rd, Zody, E., et al.\n",
            "( 2005) Genome sequence, comparative analysis and haplotype structure of the domestic dog.\n",
            "Nature 438: 803 – 819.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Lodish, H., Scott, M.\n",
            "P., Matsudaira, P., Darnell, J., Zipursky, L., Kaiser, C.\n",
            "A.,Berk, A., Krieger, M. ( 2000) Molecular cell biology  ( Freeman and Co. New York) 5th ed.. Google Scholar↵Marioni, J.\n",
            "C., Thorne, N.\n",
            "P., Tavare, S. ( 2006) Bio\n",
            "HMM: A heterogeneous hidden Markov model for segmenting array CGH data.\n",
            "Bioinformatics 22: 1144 – 1146.\n",
            "Abstract / FREE Full Text↵Mattick, J.\n",
            "S., Makunin, I.\n",
            "V. ( 2006) Non-coding RNA.\n",
            "Hum.\n",
            "Mol.\n",
            "Genet.\n",
            "15 Spec.\n",
            "No.\n",
            "1: R17 – R29.\n",
            "Google Scholar↵Mc\n",
            "Clintock, B.\n",
            "( 1929) A cytological and genetical study of triploid maize.\n",
            "Genetics 14: 180 – 222.\n",
            "FREE Full Text↵Mc\n",
            "Clintock, B.\n",
            "( 1948) Mutable loci in maize.\n",
            "Carnegie Inst.\n",
            "Wash. Year Book47: 155 – 169.\n",
            "Google Scholar↵Mendel, J.\n",
            "G. ( 1866) Verhandlungen des naturforschenden Vereines in Brünn 4 Abhandlungen, 3–47, Versuche über Pflanzenhybriden.\n",
            "Cited by Robert C. Olby (1997) on http://www.mendelweb.org/MWolby.html, accessed 2007-03-16.↵Morgan, T.\n",
            "H., Sturtevant, A.\n",
            "H., Muller, H.\n",
            "J., Bridges, C.\n",
            "B.\n",
            "( 1915) The mechanism of Mendelian heredity  ( Holt Rinehart & Winston, New York ).\n",
            "Google Scholar↵Muller, H.\n",
            "J.\n",
            "( 1927) Artificial transmutation of the gene.\n",
            "Science 46: 84 – 87.\n",
            "Cross\n",
            "Ref Google Scholar↵Nirenberg, M., Leder, P., Bernfield, M., Brimacombe, R., Trupin, J., Rottman, F.,O’Neal, C. ( 1965) RNA codewords and protein synthesis, VII.\n",
            "On the general nature of the RNA code.\n",
            "Proc.\n",
            "Natl.\n",
            "Acad.\n",
            "Sci.\n",
            "53: 1161 – 1168.\n",
            "FREE Full Text↵Ohno, S. ( 1972) in Evolution of genetic systems in Brookhaven Symposia in Biology, So much “junk” DNA in the genome, ed Smith, H.\n",
            "H..  ( Gordon & Breach, New York ), 23, pp 366 – 370.\n",
            "Medline Google Scholar↵Parra, G., Reymond, A., Dabbouseh, N., Dermitzakis, E.\n",
            "T., Castelo, R.,Thomson, T.\n",
            "M., Antonarakis, S.\n",
            "E., Guigó, R. ( 2006) Tandem chimerism as a means to increase protein complexity in the human genome.\n",
            "Genome Res.\n",
            "16: 37 – 44.\n",
            "Abstract / FREE Full Text\n",
            "Paul, J.\n",
            "( 1972) General theory of chromosome structure and gene activation in eukaryotes.\n",
            "Nature 238: 444 – 446.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Pearson, H. ( 2006) Genetics: What is a gene?\n",
            "Nature 441: 398 – 401.\n",
            "Cross\n",
            "Ref\n",
            "Medline Google Scholar↵Pedersen, J.\n",
            "S., Bejerano, G., Siepel, A., Rosenbloom, K., Lindblad-Toh, K.,Lander, E.\n",
            "S., Kent, J., Miller, W., Haussler, D. ( 2006) Identification and classification of conserved RNA secondary structures in the human genome.\n",
            "PLo\n",
            "S Comput.\n",
            "Biol.\n",
            "10.1371/journal.pcbi.0020033.\n",
            "Cross\n",
            "Ref Medline\n",
            "Google Scholar↵Ponjavic, J., Ponting, C.\n",
            "P., Lunter, G. ( 2007) Functionality or transcriptional noise?\n",
            "Evidence for selection within long noncoding RNAs.\n",
            "Genome Res.17: 556 – 565.\n",
            "Abstract / FREE Full Text↵Quelle, D.\n",
            "E., Zindy, F., Ashmun, R.\n",
            "A., Sherr, C.\n",
            "J.\n",
            "( 1995) Alternative reading frames of the INK4a tumor suppressor gene encode two unrelated proteins capable of inducing cell cycle  arrest.\n",
            "Cell 83: 993 – 1000.\n",
            "Cross\n",
            "Ref Medline\n",
            "Google Scholar↵Rheinberger, H.\n",
            "G. ( 1995) When did Darl Correns read Gregor Mendel’s paper?\n",
            "Isis 86: 612 – 616.\n",
            "Cross\n",
            "Ref Google Scholar↵Rinn, J.\n",
            "L., Euskirchen, G., Bertone, P., Martone, R., Luscombe, N.\n",
            "M.,Hartman, S., Harrison, P.\n",
            "M., Nelson, F.\n",
            "K., Mille, P., Gerstein, M., et al.\n",
            "( 2003) The transcriptional activity of human chromosome 22.\n",
            "Genes & Dev.\n",
            "17: 529 – 540.\n",
            "Abstract / FREE Full Text↵Rogic, S., Mackworth, A.\n",
            "K., Ouellette, F.\n",
            "B.\n",
            "( 2001) Evaluation of gene-finding programs on mammalian sequences.\n",
            "Genome Res.\n",
            "11: 817 – 832.\n",
            "Abstract / FREE Full Text↵Rozowsky, J., Newburger, D., Sayward, F., Wu, J., Jordan, G., Korbel, J.\n",
            "O.,Nagalakshmi, U., Yang, J., Zheng, D., Guigo, R., et al.\n",
            "( 2007) The DART classification of unannotated transcription within the ENCODE regions: Associating transcription with known and novel  loci.\n",
            "Genome Res.10.1101/gr.5696007,  (this issue).\n",
            "Abstract / FREE Full Text\n",
            "Sager, R., Kitchin, R. ( 1975) Selective silencing of eukaryotic DNA.\n",
            "Science189: 426 – 433.\n",
            "FREE Full Text↵Schadt, E.\n",
            "E., Edwards, S.\n",
            "W., Guhathakurta, D., Holder, D., Ying, L., Svetnik, V.,Leonardson, A., Hart, K.\n",
            "W., Russell, A., Li, G., et al.\n",
            "( 2004) A comprehensive transcript index of the human genome generated using microarrays and computational approaches.\n",
            "Genome Biol.\n",
            "5: R73.\n",
            "Cross\n",
            "Ref Medline\n",
            "Google Scholar↵Searls, D.\n",
            "B.\n",
            "( 1997) Abstract: Linguistic approaches to biological sequences.\n",
            "Comput.\n",
            "Appl.\n",
            "Biosci.\n",
            "13: 333 – 344.\n",
            "Abstract / FREE Full Text↵Searls, D.\n",
            "B.\n",
            "( 2001) Reading the book of life.\n",
            "Bioinformatics 17: 579 – 580.\n",
            "FREE Full Text↵Searls, D.\n",
            "B.\n",
            "( 2002) The language of genes.\n",
            "Nature 420: 211 – 217.\n",
            "Cross\n",
            "Ref\n",
            "Medline Google Scholar\n",
            "Sebat, J., Lakshmi, B., Troge, J., Alexander, J., Young, J., Lundin, P., Maner, S.,Massa, H., Walker, M., Chi, M., et al.\n",
            "( 2004) Large-scale copy number polymorphism in the human genome.\n",
            "Science 305: 525 – 528.\n",
            "Abstract / FREE Full Text↵Shi, Y., Seto, E., Chang, L.\n",
            "S., Shen, K.\n",
            "T. ( 1991) Transcriptional repression by YY1, a human GLI-Kruppel-related protein, and relief of repression by adenovirus E1A protein.\n",
            "Cell 67: 377 – 388.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Söll, D., Ohtsuka, E., Jones, D.\n",
            "S., Lohrmann, R., Hayatsu, H., Nishimura, S.,Khorana, H.\n",
            "G. ( 1965) Studies on polynucleotides, XLIX.\n",
            "Stimulation of the binding of aminoacyl-s\n",
            "RNA’s to ribosomes by ribotrinucleotides and a  survey of codon assignments for 20 amino acids.\n",
            "Proc.\n",
            "Natl.\n",
            "Acad.\n",
            "Sci.\n",
            "54: 1378 – 1385.\n",
            "FREE Full Text\n",
            "Spilianakis, C., Lalioti, M., Town, T., Lee, G., Flavell, R. ( 2005) Interchromosomal associations between alternatively expressed loci.\n",
            "Nature 435: 637 – 645.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Sturtevant, H. ( 1913) The linear arrangement of six sex-linked factors in Drosophila as shown by their mode of association.\n",
            "J. Exp.\n",
            "Zool.\n",
            "14: 43 – 59.\n",
            "Cross\n",
            "Ref Google Scholar\n",
            "Takahara, T., Kanazu, S.\n",
            "I., Yanagisawa, S., Akanuma, H. ( 2000) Heterogeneous Sp1 m\n",
            "RNAs in human Hep\n",
            "G2 cells include a product of homotypic trans -splicing.\n",
            "J. Biol.\n",
            "Chem.\n",
            "275: 38067 – 38072.\n",
            "Abstract / FREE Full Text↵Torrents, D., Suyama, M., Zdobnov, E., Bork, P. ( 2003) A genome-wide survey of human pseudogenes.\n",
            "Genome Res.\n",
            "13: 2559 – 2567.\n",
            "Abstract / FREE Full Text↵Tress, M., Martelli, P.\n",
            "L., Frankish, A., Reeves, G., Wesselink, J.\n",
            "J., Yeats, C.,Olason, P.\n",
            "I., Albrecht, M., Hegyi, H., Giorgetti, A., et al.\n",
            "( 2007) The implications of alternative splicing in the ENCODE protein complement.\n",
            "Proc.\n",
            "Natl.\n",
            "Acad.\n",
            "Sci.104: 5495 – 5500.\n",
            "Abstract / FREE Full Text↵Tschermak, E. ( 1900) Über Künstliche Kreuzung bei Pisum sativum.\n",
            "Berichte Deutsche Botanischen.\n",
            "Gesellschaft 18: 232 – 239.\n",
            "Google Scholar↵Tuan, D.\n",
            "Y., Solomon, W.\n",
            "B., London, I.\n",
            "M., Lee, D.\n",
            "P. ( 1989) An erythroid-specific, developmental-stage-independent enhancer far upstream of the human “β-like globin” genes.\n",
            "Proc.\n",
            "Natl.\n",
            "Acad.\n",
            "Sci.\n",
            "86: 2554 – 2558.\n",
            "Abstract / FREE Full Text\n",
            "Tuzun, E., Sharp, A.\n",
            "J., Bailey, J.\n",
            "A., Kaul, R., Morrison, V.\n",
            "A., Pertz, L.\n",
            "M., Haugen, E.,Hayden, H., Albertson, D., Pinkel, D., et al.\n",
            "( 2005) Fine-scale structural variation of the human genome.\n",
            "Nat.\n",
            "Genet.\n",
            "37: 727 – 732.\n",
            "Cross\n",
            "Ref Medline Google Scholar\n",
            "Vanin, E.\n",
            "F., Goldberg, G.\n",
            "I., Tucker, P.\n",
            "W., Smithies, O.\n",
            "( 1980) A mouse α-globin-related pseudogene lacking intervening sequences.\n",
            "Nature 286: 222 – 226.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Venter, J.\n",
            "C., Adams, M.\n",
            "D., Myers, E.\n",
            "W., Li, P.\n",
            "W., Mural, R.\n",
            "J., Sutton, G.\n",
            "G.,Smith, H.\n",
            "O., Yandell, M., Evans, C.\n",
            "A., Holt, R.\n",
            "A., et al.\n",
            "( 2001) The sequence of the human genome.\n",
            "Science 291: 1304 – 1351.\n",
            "Abstract / FREE Full Text\n",
            "Villa-Komaroff, L., Guttman, N., Baltimore, D., Lodishi, H.\n",
            "F. ( 1975) Complete translation of poliovirus RNA in a eukaryotic cell-free system.\n",
            "Proc.\n",
            "Natl.\n",
            "Acad.\n",
            "Sci.\n",
            "72: 4157 – 4161.\n",
            "Abstract / FREE Full Text↵Vries, H. ( 1900) Sur la loi de disjonction des hybrides.\n",
            "Comptes rendus de l’Académie des Sciences (Paris) 130: 845 – 847.\n",
            "Google Scholar↵Wade, N. ( 2003) Gene sweepstakes ends, but winner may well be wrong.\n",
            "New York Times, http://query.nytimes.com/gst/fullpage.html?sec=health&res=9A02E0D81230F930A35755C0A9659C8B63.↵Wain, H.\n",
            "M., Bruford, E.\n",
            "A., Lovering, R.\n",
            "C., Lush, M.\n",
            "J., Wright, M.\n",
            "W., Povey, S. ( 2002) Guidelines for human gene nomenclature.\n",
            "Genomics 79: 464 – 470.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Washietl, S., Hofacker, I.\n",
            "L., Lukasser, M., Huttenhofer, A., Stadler, P.\n",
            "F. ( 2005) Mapping of conserved RNA secondary structures predicts thousands of functional noncoding RNAs in the human genome.\n",
            "Nat.\n",
            "Biotechnol.23: 1383 – 1390.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Washietl, S., Pedersen, J.\n",
            "S., Korbel, J.\n",
            "O., Stocsits, C., Gruber, A.\n",
            "R.,Hackermüller, J., Hertel, J., Lindemeyer, M., Reiche, K., Tanzer, A., et al.\n",
            "( 2007) Structured RNAs in the ENCODE selected regions of the human genome.\n",
            "Genome Res.\n",
            "10.1101/gr.5650707,  (this issue).\n",
            "Abstract / FREE Full Text↵Waterston, R.\n",
            "H., Lindblad-Toh, K., Birney, E., Rogers, J., Abril, J.\n",
            "F., Agarwal, P.,Agarwala, R., Ainscough, R., Alexandersson, M., An, P., et al.\n",
            "( 2002) Initial sequencing and comparative analysis of the mouse genome.\n",
            "Nature420: 520 – 562.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Watson, J.\n",
            "D., Crick, F.\n",
            "H.\n",
            "C. ( 1953) A structure of deoxyribonucleic acid.\n",
            "Nature171: 964 – 967.\n",
            "Medline Google Scholar\n",
            "Wold, F. ( 1981) In vivo chemical modification of proteins (post-translational modification) Annu.\n",
            "Rev.\n",
            "Biochem.\n",
            "50: 783 – 814.\n",
            "Cross\n",
            "Ref Medline Google Scholar↵Yano, Y., Saito, R., Yoshida, N., Yoshiki, A., Wynshaw-Boris, A., Tomita, M.,Hirotsune, S. ( 2004) A new role for expressed pseudogenes as nc\n",
            "RNA: Regulation of m\n",
            "RNA stability of its homologous coding gene.\n",
            "J. Mol.\n",
            "Med.\n",
            "82: 414 – 422.\n",
            "Medline Google Scholar↵Zhang, Z., Harrison, P.\n",
            "M., Liu, Y., Gerstein, M. ( 2003) Millions of years of evolution preserved: A comprehensive catalog of the processed pseudogenes in the human genome.\n",
            "Genome Res.\n",
            "13: 2541 – 2558.\n",
            "Abstract / FREE Full Text↵Zhang, Z.\n",
            "D., Paccanaro, A., Fu, Y., Weissman, S., Weng, Z., Chang, J.,Snyder, M., Gerstein, M.\n",
            "B.\n",
            "( 2007) Statistical analysis of the genomic distribution and correlation of regulatory elements in the ENCODE regions.\n",
            "Genome Res.10.1101/gr.5573107,  (this issue).\n",
            "Abstract / FREE Full Text↵Zheng, D., Gerstein, M.\n",
            "B.\n",
            "( 2007) The ambiguous boundary between genes and pseudogenes: The dead rise up, or do they?\n",
            "Trends Genet.\n",
            "23: 219 – 224.\n",
            "Cross\n",
            "Ref\n",
            "Medline Google Scholar↵Zheng, D., Zhang, Z., Harrison, P.\n",
            "M., Karro, J., Carriero, N., Gerstein, M. ( 2005) Integrated pseudogene annotation for human chromosome 22: Evidence for transcription.\n",
            "J. Mol.\n",
            "Biol.\n",
            "349: 27 – 45.\n",
            "Cross\n",
            "Ref Medline Google Scholar\n",
            "Zheng, D., Frankish, A., Baertsch, R., Kapranov, P., Reymond, A., Choo, S.\n",
            "W.,Lu, Y., Denoeud, F., Antonarakis, S.\n",
            "E., Snyder, M., et al.\n",
            "( 2007) Pseudogenes in the ENCODE regions: Consensus annotation, analysis of transcription, and evolution.\n",
            "Genome Res.\n",
            "10.1101/gr.5586307,  (this issue).\n",
            "Abstract / FREE Full Text\n",
            "Cite\n",
            "ULike Delicious Digg\n",
            "Facebook\n",
            "Google+ Reddit Twitter\n",
            "What's this?\n",
            "Related Articles\n",
            "PERSPECTIVE:\n",
            "Raising the estimate of functional human sequences\n",
            "Michael Pheasant and John S. Mattick\n",
            "Genome Res.\n",
            "September 2007 17: 1245 - 1253; Published in Advance August 9, 2007, doi: 10.1101/gr.6406307Abstract Full Text Full Text (PDF)PERSPECTIVE:\n",
            "Origin of phenotypes: Genes and transcripts\n",
            "Thomas R. Gingeras\n",
            "Genome Res.\n",
            "June 2007 17: 682 - 690; doi: 10.1101/gr.6525007Abstract Full Text Full Text (PDF)OPEN ACCESS ARTICLE« Previous | Next Article »Table of Contents\n",
            "OPEN ACCESS ARTICLEThis Articledoi: 10.1101/gr.6339607Genome Res.\n",
            "2007.\n",
            "17: 669-681Copyright © 2007, Cold Spring Harbor Laboratory Press\n",
            "Abstract Free» Full Text Free\n",
            "Full Text (PDF) Free\n",
            "Fig._1_Timeline_Poster.pdf- Article Category\n",
            "PERSPECTIVE+ Services+ Citing Articles+ Google Scholar+ Pub\n",
            "Med/NCBI- Related Articles\n",
            "Raising the estimate of functional human sequences\n",
            "Pheasant, Michael et al.\n",
            "Origin of phenotypes: Genes and transcripts\n",
            "Gingeras, Thomas R. et al.+ Share- Metrics\n",
            "Recent Updates\n",
            "Tweets by @genomeresearch+ Most Read Articles\n",
            "View all ...\n",
            "Navigate This Article\n",
            "Top\n",
            "Abstract\n",
            "Introduction\n",
            "History of the gene, 1860 to just before ENCODEProblematic issues with the current definition of a gene\n",
            "The modern ENCODE view of dispersed genome activity\n",
            "The importance of gene models for interpreting the high-throughput experiment in ENCODEToward an updated definition of a gene\n",
            "Conclusion: What is function?\n",
            "Acknowledgments\n",
            "Footnotes\n",
            "References\n",
            "Current Issue\n",
            "March 2018, 28 (3)+ From the Cover\n",
            "Alert me to new issues of Genome Research\n",
            "Advance Online Articles\n",
            "Submit a Manuscript\n",
            "GR in the News\n",
            "Editorial Board\n",
            "Permissions\n",
            "E-mail Alerts & RSS Feeds\n",
            "Recommend to Your Library\n",
            "Job Opportunities\n",
            "Home About Archive Submit Subscribe Advertise Author Info Contact Help\n",
            "Copyright © 2018 by Cold Spring Harbor Laboratory Press\n",
            "Print ISSN: 1088-9051Online ISSN: 1549-5469\n",
            "\n",
            "\n"
          ]
        }
      ],
      "source": [
        "target_doc_id = \"D1806793\"\n",
        "\n",
        "doc2 = \"\"\n",
        "\n",
        "for doc in dataset.docs_iter():\n",
        "    if doc.doc_id == target_doc_id:\n",
        "        doc2 = doc.body\n",
        "        break\n",
        "\n",
        "if doc2 is not None:\n",
        "    print(f\"Body of Document ID {target_doc_id}:\\n{doc2}\")\n",
        "else:\n",
        "    print(f\"No document found with ID {target_doc_id}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "eBgwD4ULom_2"
      },
      "outputs": [],
      "source": [
        "ps = PorterStemmer()\n",
        "stop_words = set(stopwords.words('english'))\n",
        "\n",
        "def preprocess_string(input_string):\n",
        "\n",
        "    words = word_tokenize(input_string)\n",
        "\n",
        "    processed_words = [ps.stem(word) for word in words if word.lower() not in stop_words]\n",
        "\n",
        "    return ' '.join(processed_words)\n",
        "\n",
        "\n",
        "# Preprocess the strings\n",
        "doc1 = preprocess_string(doc1)\n",
        "doc2 = preprocess_string(doc2)\n",
        "query_text = preprocess_string(query_text)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Applying Selected Axioms to see which doc is preffered\n",
        "\n",
        "\n",
        "*   TFC1: assigns higher scores\n",
        "to documents that contain a query term more often normalised by document length\n",
        "\n",
        "*   LB1:If there is a query term t with tf(t, d1) = 0\n",
        "and tf(t, d2) > 0, then d1 < d2.\n",
        "\n",
        "*   LNC1:  two documents are compared that have the same term frequency\n",
        "for all query terms (10%-softening). Then the shorter document is preferred.\n",
        "\n",
        "*   AND: from a document pair d1 and d2 where only d1 contains all\n",
        "query terms, d1 is preferred.\n",
        "\n",
        "*   PROX1-3: The closer the query terms appear to each other in a document, the more relevant the document is.\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "Dx7lz5kyT2eP"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 81
        },
        "id": "y73Md5pcDHMY",
        "outputId": "e28184b0-13d1-40db-bac4-d43acb36e13b"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                 Query                 Document 1                 Document 2  \\\n",
              "0  exon definit biolog  definit refer ( + imag )   skip main page content ho   \n",
              "\n",
              "   TFC1  LB1  LNC1  AND   PROX1+PROX2+PROX3   \n",
              "0     1    0     1    1                    0  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-973e2fd2-6871-483b-8807-02feae991be3\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Query</th>\n",
              "      <th>Document 1</th>\n",
              "      <th>Document 2</th>\n",
              "      <th>TFC1</th>\n",
              "      <th>LB1</th>\n",
              "      <th>LNC1</th>\n",
              "      <th>AND</th>\n",
              "      <th>PROX1+PROX2+PROX3</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>exon definit biolog</td>\n",
              "      <td>definit refer ( + imag )</td>\n",
              "      <td>skip main page content ho</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-973e2fd2-6871-483b-8807-02feae991be3')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-973e2fd2-6871-483b-8807-02feae991be3 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-973e2fd2-6871-483b-8807-02feae991be3');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"pairwise_instance\",\n  \"rows\": 1,\n  \"fields\": [\n    {\n      \"column\": \"Query\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"exon definit biolog\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Document 1\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"definit refer ( + imag ) \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Document 2\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"skip main page content ho\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"TFC1\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 1,\n        \"max\": 1,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"LB1\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 0,\n        \"max\": 0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"LNC1\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 1,\n        \"max\": 1,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"AND\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 1,\n        \"max\": 1,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \" PROX1+PROX2+PROX3 \",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 0,\n        \"max\": 0,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 22
        }
      ],
      "source": [
        "index_path = \"\"\n",
        "\n",
        "# Create an instance of the pairwise class\n",
        "pairwise_instance = pairwise(query_text, doc1, doc2, index_path)\n",
        "\n",
        "axiom_classes = [\"TFC1\", \"LB1\", \"LNC1\", \"AND\", \" PROX1+PROX2+PROX3 \"]\n",
        "\n",
        "pairwise_instance.explain(query_text, doc1, doc2, axiom_classes)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "More details on the following axiom -\n",
        "\n",
        "\n",
        "*   TFC1\n",
        "*   LNC1\n",
        "\n"
      ],
      "metadata": {
        "id": "49HT261NUZFD"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Hj1bERQt3HLU",
        "outputId": "347900b9-1083-4160-f20d-c736f58fcdbd"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Normalised tf score in doc1: 0.014354066985645933\n",
            "Normalised tf score in doc2: 0.0012175324675324675\n",
            "None\n"
          ]
        }
      ],
      "source": [
        "axiom_name = \"TFC1\"\n",
        "explanation = pairwise_instance.explain_details(query_text, doc1, doc2, axiom_name)\n",
        "print(explanation)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "R1thMHbP6GBR",
        "outputId": "5813ea25-0f59-41dc-899e-0cc18927136d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of query terms document 1: 3\n",
            "Number of query terms document 2: 3\n",
            "\n",
            "Length of document1: 3806\n",
            "Length of document2: 62188\n",
            "None\n"
          ]
        }
      ],
      "source": [
        "axiom_name = \"LNC1\"\n",
        "explanation = pairwise_instance.explain_details(query_text, doc1, doc2, axiom_name)\n",
        "print(explanation)"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "5vgUcZgHUX7e"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}